<h3>Genetic Value Analysis</h3>
<p>
	The genetic value analysis proceeds according to the following steps:
</p>

<ol type="1">
	<li style="padding-bottom:15px">
		First, the average of all pairwise <b>kinship coefficients</b> are
		calculated for each living animal in the breeding colony relative
		to every other living animal in the colony according, to the algorithm
		of Lange (2002). The grand mean kinship in the breeding colony is then
		calculated from the distribution of individual mean kinships, along with
		other summary statistics describing this distribution.
	</li>
	<li style="padding-bottom:15px">
		<p>
		Second, measures of individual mean kinship for each animal are then
		calculated in terms of standard deviates from this grand mean kinship,
		as <em>z = (x-&mu;)/&sigma;</em>, i.e., "<b>z-scores</b>", in which x is the estimate of
		individual mean kinship, &mu; is the grand mean kinship, and &sigma; is the
		standard deviation of the distribution of all individual mean kinship
		values of living animals in the colony.
		</p>
		<p>
		These z-scores indicate the extent and direction to which each animal's
		individual mean kinship deviates from the grand mean kinship in the colony.
		An animal having a positive z-score has individual mean kinship that lies
		above the colony grand mean kinship, suggesting that this animal's genome
		is over-represented in the colony, making the animal of lesser genetic value
		than others with negative z-scores.  Conversely, an animal having a negative
		z-score has individual mean kinship that lies below the grand mean kinship,
		suggesting that this animal's genome is underrepresented in the colony, making
		the animal of greater genetic value than animals with positive z-scores.
		</p>
	</li>
	<li style="padding-bottom:15px">
		Third, <b>genome uniqueness</b> values are calculated using a gene-drop simulation
		according to MacCluer et al. (1986) and Ballou & Lacy (1995), by assigning unique
		alleles to all pedigree founders, and simulating their segregation throughout the
		pedigree according to Mendelian rules.  Genome uniqueness is a measure of the
		probability that an animal possesses founder alleles that are present in at most
		<em>x</em> other animals (usually 0-3), and thus are rare and at risk of being lost from
		the population.  A range of 2 to 100,000 simulations may be selected.  A minimum
		of 1,000 simulations is recommended.  A genome uniqueness threshold value between
		0-3 should also be selected, as desired.
	</li>
	<li style="padding-bottom:15px">
		Following the calculation of mean kinship, related z-scores, and genome uniqueness
		for all animals, animals are ranked from the highest genetic value (ranking = 1)
		to the lowest, according to the following rules:
		<ol type=i>
			<li style="padding-top:5px; padding-bottom:5px">
				Imported animals without offspring are automatically ranked highest in
				genetic value, followed by
			</li>
			<li style="padding-bottom:5px">
				All animals with greater than 10% genome uniqueness, ordered first by
				genome uniqueness from largest to smallest values, and by z-score within
				the same genome uniqueness from smallest to largest values; followed by
			</li>
			<li style="padding-bottom:5px">
				All remaining animals (those with = 10% genome uniqueness), ranked by
				z-score from smallest to largest values; followed by
			</li>
			<li>
				A threshold drawn at those animals with genome uniqueness = 10% and
				z-scores = 0.25.  Animals falling below this threshold are indicated by
				highlighting in the results.  Animals ranking above this threshold are
				considered to be of high genetic value, while animals that rank below this
				threshold are not recommended for breeding, but may be suitable for research
				assignment or sale.
			</li>
		</ol>
	</li>
</ol>
<br>
<h3>Breeding Group Formation</h3>
<ul style="list-style-type:disc">
	<li style="padding-bottom: 15px">
		In large colonies, most or all genetic management may take
		place primarily during development of the housing/breeding
		group. A genetic value analysis may be used as the basis
		for creating potential new breeding groups that will maximize
		genetic diversity within the colony as a whole.
	</li>
	<li style="padding-bottom: 15px">
		When using the results of a genetic value analysis to develop
		new breeding groups, it is important to recognize that animals
		of high genetic value may still be closely related to other high-value
		animals, and that genetic diversity in the colony can be greatly
		reduced if relatives are allowed to produce viable offspring.
	</li>
	<li style="padding-bottom: 15px">
		Following a genetic value analysis of all breeding colony animals,
		only animals lying above the threshold will be considered for potential
		new breeding groups.  In this analysis, potential new breeding groups
		are formed based on relationships that must be excluded, the threshold
		value of within-group relatedness that may be tolerated, relationships
		that can be safely ignored, and any additional specific group criteria.
		For example, relatedness greater than second cousin between breeding males,
		and between breeding males and females should be of concern; however,
		relatedness between breeding adults and animals &le; 1 year of age, or between
		females of any age might be ignored, depending on the social group structure desired.
	</li>
	<li style="padding-bottom: 15px">
		This algorithm will assign randomly-selected animals to hypothetical groups,
		with the constraint that they cannot be related at the specified threshold
		level or greater to animals already assigned to the group.  This process is
		repeated 10,000 times to thoroughly sample the collection of possible groups
		that can be formed from the high-value set of candidate animals being considered.
		Upon completion, this algorithm will return the best set of groups, defined as
		the set with the largest minimum group size.
	</li>
</ul>
