<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Gene drop simulation based on the provided pedigree information — geneDrop • nprcgenekeepr</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Gene drop simulation based on the provided pedigree information — geneDrop"><meta name="description" content="Part of Genetic Value Analysis"><meta property="og:description" content="Part of Genetic Value Analysis"><meta property="og:image" content="https://github.com/rmsharp/nprcgenekeepr/reference/figures/card.png"><meta property="og:image:alt" content="Conserve Genes"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">nprcgenekeepr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.6</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/a2interactive.html">Interactive Use of nprcgenekeepr</a></li>
    <li><a class="dropdown-item" href="../articles/a3manual.html">Genetic Management Tools Manual</a></li>
    <li><a class="dropdown-item" href="../articles/ColonyManagerTutorial.html">GeneKeepR: A Colony Manager's Tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/shiny_app_use.html">GeneKeepR Tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/simulatedKValues.html">Quick Example of Simulated Kinships with Partial Parentage</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/rmsharp/nprcgenekeepr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Gene drop simulation based on the provided pedigree information</h1>
      <small class="dont-index">Source: <a href="https://github.com/rmsharp/nprcgenekeepr/blob/master/R/geneDrop.R"><code>R/geneDrop.R</code></a></small>
      <div class="d-none name"><code>geneDrop.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Part of Genetic Value Analysis</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">geneDrop</span><span class="op">(</span></span>
<span>  <span class="va">ids</span>,</span>
<span>  <span class="va">sires</span>,</span>
<span>  <span class="va">dams</span>,</span>
<span>  <span class="va">gen</span>,</span>
<span>  genotype <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  n <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>  updateProgress <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-ids">ids<a class="anchor" aria-label="anchor" href="#arg-ids"></a></dt>
<dd><p>A character vector of IDs for a set of animals.</p></dd>


<dt id="arg-sires">sires<a class="anchor" aria-label="anchor" href="#arg-sires"></a></dt>
<dd><p>A character vector with IDS of the sires for the set of
animals. <code>NA</code> is used for missing sires.</p></dd>


<dt id="arg-dams">dams<a class="anchor" aria-label="anchor" href="#arg-dams"></a></dt>
<dd><p>A character vector with IDS of the dams for the set of
animals. <code>NA</code> is used for missing dams.</p></dd>


<dt id="arg-gen">gen<a class="anchor" aria-label="anchor" href="#arg-gen"></a></dt>
<dd><p>An integer vector indicating the generation number for each
animal.</p></dd>


<dt id="arg-genotype">genotype<a class="anchor" aria-label="anchor" href="#arg-genotype"></a></dt>
<dd><p>A dataframe containing known genotypes. It has three
columns:  <code>id</code>, <code>first</code>, and <code>second</code>. The second and third
columns contain the integers indicating the observed genotypes.</p>
<p>The gene dropping method from <em>Pedigree analysis by computer simulation</em>
by Jean W MacCluer, John L Vandeberg, and Oliver A Ryder (1986)
&lt;doi:10.1002/zoo.1430050209&gt; is used in the genetic value calculations.</p>
<p>Currently there is no means of handling knowing only one haplotype.
It will be easy to add another column to handle situations where only one
allele is observed and it is not known to be homozygous or heterozygous. The
new fourth column could have a frequency for homozygosity that could be
used in the gene dropping algorithm.</p>
<p>The genotypes are using indirection (integer instead of character) to
indicate the genes because the manipulation of character strings was found
to take 20-35 times longer to perform.</p>
<p>Adding additional columns to <code>genotype</code> does not significantly affect
the time require. Thus, it is convenient to add the corresponding haplotype
names to the dataframe using <code>first_name</code> and <code>second_name</code>.</p></dd>


<dt id="arg-n">n<a class="anchor" aria-label="anchor" href="#arg-n"></a></dt>
<dd><p>integer indicating the number of iterations to simulate.
Default is 5000.</p></dd>


<dt id="arg-updateprogress">updateProgress<a class="anchor" aria-label="anchor" href="#arg-updateprogress"></a></dt>
<dd><p>function or NULL. If this function is defined, it
will be called during each iteration to update a
<code><a href="https://rdrr.io/pkg/shiny/man/Progress.html" class="external-link">shiny::Progress</a></code> object.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data.frame <code>id, parent, V1 ... Vn</code>
A data.frame providing the maternal and paternal alleles for an animal
for each iteration. The first two columns provide the animal's ID and
whether the allele came from the sire or dam. These are followed by
<code>n</code> columns indicating the allele for that iteration.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="co">## We usually defined `n` to be &gt;= 5000</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rmsharp.github.io/nprcgenekeepr/" class="external-link">nprcgenekeepr</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">ped</span> <span class="op">&lt;-</span> <span class="fu">nprcgenekeepr</span><span class="fu">::</span><span class="va"><a href="lacy1989Ped.html">lacy1989Ped</a></span></span></span>
<span class="r-in"><span><span class="va">allelesNew</span> <span class="op">&lt;-</span> <span class="fu">geneDrop</span><span class="op">(</span><span class="va">ped</span><span class="op">$</span><span class="va">id</span>, <span class="va">ped</span><span class="op">$</span><span class="va">sire</span>, <span class="va">ped</span><span class="op">$</span><span class="va">dam</span>, <span class="va">ped</span><span class="op">$</span><span class="va">gen</span>,</span></span>
<span class="r-in"><span>                      genotype <span class="op">=</span> <span class="cn">NULL</span>, n <span class="op">=</span> <span class="fl">50</span>, updateProgress <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">genotype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>id <span class="op">=</span> <span class="va">ped</span><span class="op">$</span><span class="va">id</span>,</span></span>
<span class="r-in"><span>                       first_allele <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="cn">NA</span>, <span class="st">"A001_B001"</span>, <span class="st">"A001_B002"</span>,</span></span>
<span class="r-in"><span>                                        <span class="cn">NA</span>, <span class="st">"A001_B002"</span>, <span class="st">"A001_B001"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                       second_allele <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="cn">NA</span>, <span class="st">"A010_B001"</span>, <span class="st">"A001_B001"</span>,</span></span>
<span class="r-in"><span>                                         <span class="cn">NA</span>, <span class="cn">NA</span>, <span class="cn">NA</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                       stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">pedWithGenotype</span> <span class="op">&lt;-</span> <span class="fu"><a href="addGenotype.html">addGenotype</a></span><span class="op">(</span><span class="va">ped</span>, <span class="va">genotype</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">pedGenotype</span> <span class="op">&lt;-</span> <span class="fu"><a href="getGVGenotype.html">getGVGenotype</a></span><span class="op">(</span><span class="va">pedWithGenotype</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">allelesNewGen</span> <span class="op">&lt;-</span> <span class="fu">geneDrop</span><span class="op">(</span><span class="va">ped</span><span class="op">$</span><span class="va">id</span>, <span class="va">ped</span><span class="op">$</span><span class="va">sire</span>, <span class="va">ped</span><span class="op">$</span><span class="va">dam</span>, <span class="va">ped</span><span class="op">$</span><span class="va">gen</span>,</span></span>
<span class="r-in"><span>                         genotype <span class="op">=</span> <span class="va">pedGenotype</span>,</span></span>
<span class="r-in"><span>                         n <span class="op">=</span> <span class="fl">5</span>, updateProgress <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Michael Raboin, Terry Therneau, Amanda Vinson, R. Mark Sharp, Matthew Schultz, Southwest National Primate Research Center NIH grant P51 RR13986, Oregon National Primate Research Center grant P51 OD011092.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

