<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ nprcgenekeepr</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><meta property="og:image" content="https://github.com/rmsharp/nprcgenekeepr/reference/figures/card.png"><meta property="og:image:alt" content="Conserve Genes"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">nprcgenekeepr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.6</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/a2interactive.html">Interactive Use of nprcgenekeepr</a></li>
    <li><a class="dropdown-item" href="../articles/a3manual.html">Genetic Management Tools Manual</a></li>
    <li><a class="dropdown-item" href="../articles/ColonyManagerTutorial.html">GeneKeepR: A Colony Manager's Tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/shiny_app_use.html">GeneKeepR Tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/simulatedKValues.html">Quick Example of Simulated Kinships with Partial Parentage</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/rmsharp/nprcgenekeepr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="addAnimalsWithNoRelative.html">addAnimalsWithNoRelative()</a></code>

        </dt>
        <dd>Adds an NA value for all animals without a relative</dd>
      </dl><dl><dt>

          <code><a href="addBackSecondParents.html">addBackSecondParents()</a></code>

        </dt>
        <dd>Add back single parents trimmed pedigree</dd>
      </dl><dl><dt>

          <code><a href="addErrTxt.html">addErrTxt()</a></code>

        </dt>
        <dd>Concatenates any errors from nprcgenekeeprErr into narrative form</dd>
      </dl><dl><dt>

          <code><a href="addGenotype.html">addGenotype()</a></code>

        </dt>
        <dd>Add genotype data to pedigree file</dd>
      </dl><dl><dt>

          <code><a href="addGroupOfUnusedAnimals.html">addGroupOfUnusedAnimals()</a></code>

        </dt>
        <dd>addGroupOfUnusedAnimals adds a group to the saved groups if needed</dd>
      </dl><dl><dt>

          <code><a href="addIdRecords.html">addIdRecords()</a></code>

        </dt>
        <dd>addIdRecords Adds Ego records added having NAs for parent IDs</dd>
      </dl><dl><dt>

          <code><a href="addKinshipValueCount.html">addKinshipValueCount()</a></code>

        </dt>
        <dd>Add to count the number of occurrences of a kinship value seen for a pair of individuals in one or more simulated pedigrees.</dd>
      </dl><dl><dt>

          <code><a href="addParents.html">addParents()</a></code>

        </dt>
        <dd>Add parents</dd>
      </dl><dl><dt>

          <code><a href="addSexAndAgeToGroup.html">addSexAndAgeToGroup()</a></code>

        </dt>
        <dd>Forms a dataframe with Id, Sex, and current Age given a list of Ids and a pedigree</dd>
      </dl><dl><dt>

          <code><a href="addUIds.html">addUIds()</a></code>

        </dt>
        <dd>Eliminates partial parentage situations by adding unique placeholder IDs for the unknown parent.</dd>
      </dl><dl><dt>

          <code><a href="agePyramidPlot.html">agePyramidPlot()</a></code>

        </dt>
        <dd>Form age pyramid plot</dd>
      </dl><dl><dt>

          <code><a href="allTrueNoNA.html">allTrueNoNA()</a></code>

        </dt>
        <dd>Returns <code>TRUE</code> if every member of the vector is <code>TRUE</code>.</dd>
      </dl><dl><dt>

          <code><a href="alleleFreq.html">alleleFreq()</a></code>

        </dt>
        <dd>Calculates the count of each allele in the provided vector.</dd>
      </dl><dl><dt>

          <code><a href="assignAlleles.html">assignAlleles()</a></code>

        </dt>
        <dd>Assign parent alleles randomly</dd>
      </dl><dl><dt>

          <code><a href="calcA.html">calcA()</a></code>

        </dt>
        <dd>Calculates <code>a</code>, the number of an individual's alleles that are rare in each simulation.</dd>
      </dl><dl><dt>

          <code><a href="calcAge.html">calcAge()</a></code>

        </dt>
        <dd>Calculate animal ages.</dd>
      </dl><dl><dt>

          <code><a href="calcFE.html">calcFE()</a></code>

        </dt>
        <dd>Calculates founder Equivalents</dd>
      </dl><dl><dt>

          <code><a href="calcFEFG.html">calcFEFG()</a></code>

        </dt>
        <dd>Calculates Founder Equivalents and Founder Genome Equivalents</dd>
      </dl><dl><dt>

          <code><a href="calcFG.html">calcFG()</a></code>

        </dt>
        <dd>Calculates Founder Genome Equivalents</dd>
      </dl><dl><dt>

          <code><a href="calcGU.html">calcGU()</a></code>

        </dt>
        <dd>Calculates genome uniqueness for each ID that is part of the population.</dd>
      </dl><dl><dt>

          <code><a href="calcRetention.html">calcRetention()</a></code>

        </dt>
        <dd>Calculates Allelic Retention</dd>
      </dl><dl><dt>

          <code><a href="calculateSexRatio.html">calculateSexRatio()</a></code>

        </dt>
        <dd>Calculates the sex ratio (number of non-males / number of males) given animal Ids and their pedigree</dd>
      </dl><dl><dt>

          <code><a href="checkChangedColAndErrorLst.html">checkChangedColAndErrorLst()</a></code>

        </dt>
        <dd>checkChangedColAndErrorLst examines errorLst for errors and errorLst$changeCols non-empty fields</dd>
      </dl><dl><dt>

          <code><a href="checkChangedColsLst.html">checkChangedColsLst()</a></code>

        </dt>
        <dd>checkChangedColsLst examines list for non-empty fields</dd>
      </dl><dl><dt>

          <code><a href="checkErrorLst.html">checkErrorLst()</a></code>

        </dt>
        <dd>checkErrorLst examines list for non-empty fields</dd>
      </dl><dl><dt>

          <code><a href="checkGenotypeFile.html">checkGenotypeFile()</a></code>

        </dt>
        <dd>Check genotype file</dd>
      </dl><dl><dt>

          <code><a href="checkParentAge.html">checkParentAge()</a></code>

        </dt>
        <dd>Check parent ages to be at least <code>minParentAge</code></dd>
      </dl><dl><dt>

          <code><a href="checkRequiredCols.html">checkRequiredCols()</a></code>

        </dt>
        <dd>Examines column names, <code>cols</code> for required column names</dd>
      </dl><dl><dt>

          <code><a href="chooseAlleles.html">chooseAlleles()</a></code>

        </dt>
        <dd>Combines two vectors of alleles by randomly selecting one allele or the other at each position.</dd>
      </dl><dl><dt>

          <code><a href="chooseAllelesChar.html">chooseAllelesChar()</a></code>

        </dt>
        <dd>Combines two vectors of alleles when alleles are character vectors.</dd>
      </dl><dl><dt>

          <code><a href="chooseDate.html">chooseDate()</a></code>

        </dt>
        <dd>Choose date based on <code>earlier</code> flag.</dd>
      </dl><dl><dt>

          <code><a href="colChange.html">colChange()</a></code>

        </dt>
        <dd>colChange internal function to describe column names transformation</dd>
      </dl><dl><dt>

          <code><a href="convertAncestry.html">convertAncestry()</a></code>

        </dt>
        <dd>Converts the ancestry information to a standardized code</dd>
      </dl><dl><dt>

          <code><a href="convertDate.html">convertDate()</a></code>

        </dt>
        <dd>Converts date columns formatted as characters to be of type datetime</dd>
      </dl><dl><dt>

          <code><a href="convertFromCenter.html">convertFromCenter()</a></code>

        </dt>
        <dd>Converts the fromCenter information to a standardized code</dd>
      </dl><dl><dt>

          <code><a href="convertRelationships.html">convertRelationships()</a></code>

        </dt>
        <dd>Converts pairwise kinship values to a relationship category descriptor.</dd>
      </dl><dl><dt>

          <code><a href="convertSexCodes.html">convertSexCodes()</a></code>

        </dt>
        <dd>Converts sex indicator for an individual to a standardized codes.</dd>
      </dl><dl><dt>

          <code><a href="convertStatusCodes.html">convertStatusCodes()</a></code>

        </dt>
        <dd>Converts status indicators to a Standardized code</dd>
      </dl><dl><dt>

          <code><a href="correctParentSex.html">correctParentSex()</a></code>

        </dt>
        <dd>Sets sex for animals listed as either a sire or dam.</dd>
      </dl><dl><dt>

          <code><a href="countFirstOrder.html">countFirstOrder()</a></code>

        </dt>
        <dd>Count first-order relatives.</dd>
      </dl><dl><dt>

          <code><a href="countKinshipValues.html">countKinshipValues()</a></code>

        </dt>
        <dd>Counts the number of occurrences of each kinship value seen for a pair of individuals in a series of simulated pedigrees.</dd>
      </dl><dl><dt>

          <code><a href="countLoops.html">countLoops()</a></code>

        </dt>
        <dd>Count the number of loops in a pedigree tree.</dd>
      </dl><dl><dt>

          <code><a href="createExampleFiles.html">createExampleFiles()</a></code>

        </dt>
        <dd>Creates a folder with CSV files containing example pedigrees and ID lists used to demonstrate the package.</dd>
      </dl><dl><dt>

          <code><a href="createPedOne.html">createPedOne()</a></code>

        </dt>
        <dd>createPedOne makes the pedOne data object</dd>
      </dl><dl><dt>

          <code><a href="createPedSix.html">createPedSix()</a></code>

        </dt>
        <dd>createPedSix makes the pedSix data object</dd>
      </dl><dl><dt>

          <code><a href="createPedTree.html">createPedTree()</a></code>

        </dt>
        <dd>Create a pedigree tree (PedTree).</dd>
      </dl><dl><dt>

          <code><a href="createSimKinships.html">createSimKinships()</a></code>

        </dt>
        <dd>Makes a list object of kinship matrices from simulated pedigrees of possible parents for animals with unknown parents</dd>
      </dl><dl><dt>

          <code><a href="create_wkbk.html">create_wkbk()</a></code>

        </dt>
        <dd>Creates an Excel workbook with worksheets.</dd>
      </dl><dl><dt>

          <code><a href="cumulateSimKinships.html">cumulateSimKinships()</a></code>

        </dt>
        <dd>Makes a list object containing kinship summary statistics using the list object from __createSimKinships__.</dd>
      </dl><dl><dt>

          <code><a href="dataframe2string.html">dataframe2string()</a></code>

        </dt>
        <dd>dataframe2string converts a data.frame object to a character vector</dd>
      </dl><dl><dt>

          <code><a href="exampleNprcgenekeeprConfig.html">exampleNprcgenekeeprConfig</a></code>

        </dt>
        <dd>exampleNprcgenekeeprConfig is a loadable version of the example configuration file <code>example_nprcgenekeepr_config</code></dd>
      </dl><dl><dt>

          <code><a href="examplePedigree.html">examplePedigree</a></code>

        </dt>
        <dd>examplePedigree is a pedigree object created by <code>qcStudbook</code></dd>
      </dl><dl><dt>

          <code><a href="fillBins.html">fillBins()</a></code>

        </dt>
        <dd>fillBins Fill bins represented by list of two lists <code>males</code> and <code>females</code>.</dd>
      </dl><dl><dt>

          <code><a href="fillGroupMembers.html">fillGroupMembers()</a></code>

        </dt>
        <dd>Forms and fills list of animals groups based on provided constraints</dd>
      </dl><dl><dt>

          <code><a href="fillGroupMembersWithSexRatio.html">fillGroupMembersWithSexRatio()</a></code>

        </dt>
        <dd>Forms breeding group(s) with an effort to match a specified sex ratio</dd>
      </dl><dl><dt>

          <code><a href="filterAge.html">filterAge()</a></code>

        </dt>
        <dd>Removes kinship values where an animal is less than the minAge</dd>
      </dl><dl><dt>

          <code><a href="filterKinMatrix.html">filterKinMatrix()</a></code>

        </dt>
        <dd>Filters a kinship matrix to include only the egos listed in 'ids'</dd>
      </dl><dl><dt>

          <code><a href="filterPairs.html">filterPairs()</a></code>

        </dt>
        <dd>Filters kinship values from a long-format kinship table based on the sexes of the two animals involved.</dd>
      </dl><dl><dt>

          <code><a href="filterReport.html">filterReport()</a></code>

        </dt>
        <dd>Filters a genetic value report down to only the specified animals</dd>
      </dl><dl><dt>

          <code><a href="filterThreshold.html">filterThreshold()</a></code>

        </dt>
        <dd>Filters kinship to remove rows with kinship values less than the specified threshold</dd>
      </dl><dl><dt>

          <code><a href="finalRpt.html">finalRpt</a></code>

        </dt>
        <dd>finalRpt is a list object created from the list object <em>rpt</em> prepared by <code>reportGV</code>. It is created inside <code>orderReport</code>. This version is at the state just prior to calling <code>rankSubjects</code> inside <code>orderReport</code>.</dd>
      </dl><dl><dt>

          <code><a href="findGeneration.html">findGeneration()</a></code>

        </dt>
        <dd>Determines the generation number for each id.</dd>
      </dl><dl><dt>

          <code><a href="findLoops.html">findLoops()</a></code>

        </dt>
        <dd>Find loops in a pedigree tree</dd>
      </dl><dl><dt>

          <code><a href="findOffspring.html">findOffspring()</a></code>

        </dt>
        <dd>Finds the number of total offspring for each animal in the provided pedigree.</dd>
      </dl><dl><dt>

          <code><a href="findPedigreeNumber.html">findPedigreeNumber()</a></code>

        </dt>
        <dd>Determines the generation number for each id.</dd>
      </dl><dl><dt>

          <code><a href="fixColumnNames.html">fixColumnNames()</a></code>

        </dt>
        <dd>fixColumnNames changes original column names and into standardized names.</dd>
      </dl><dl><dt>

          <code><a href="fixGenotypeCols.html">fixGenotypeCols()</a></code>

        </dt>
        <dd>Reformat names of observed genotype columns</dd>
      </dl><dl><dt>

          <code><a href="focalAnimals.html">focalAnimals</a></code>

        </dt>
        <dd>focalAnimals is a dataframe with one column (_id_) containing the of animal Ids from the __examplePedigree__ pedigree.</dd>
      </dl><dl><dt>

          <code><a href="geneDrop.html">geneDrop()</a></code>

        </dt>
        <dd>Gene drop simulation based on the provided pedigree information</dd>
      </dl><dl><dt>

          <code><a href="getAncestors.html">getAncestors()</a></code>

        </dt>
        <dd>Recursively create a character vector of ancestors for an individual ID.</dd>
      </dl><dl><dt>

          <code><a href="getAnimalsWithHighKinship.html">getAnimalsWithHighKinship()</a></code>

        </dt>
        <dd>Forms a list of animal Ids and animals related to them</dd>
      </dl><dl><dt>

          <code><a href="getChangedColsTab.html">getChangedColsTab()</a></code>

        </dt>
        <dd>getChangedColsTab skeleton of list of errors</dd>
      </dl><dl><dt>

          <code><a href="getConfigFileName.html">getConfigFileName()</a></code>

        </dt>
        <dd>getConfigFileName returns the configuration file name appropriate for the system.</dd>
      </dl><dl><dt>

          <code><a href="getCurrentAge.html">getCurrentAge()</a></code>

        </dt>
        <dd>Age in years using the provided birthdate.</dd>
      </dl><dl><dt>

          <code><a href="getDateColNames.html">getDateColNames()</a></code>

        </dt>
        <dd>Vector of date column names</dd>
      </dl><dl><dt>

          <code><a href="getDateErrorsAndConvertDatesInPed.html">getDateErrorsAndConvertDatesInPed()</a></code>

        </dt>
        <dd>Converts columns of dates in text form to <code>Date</code> object columns</dd>
      </dl><dl><dt>

          <code><a href="getDatedFilename.html">getDatedFilename()</a></code>

        </dt>
        <dd>Returns a character vector with an file name having the date prepended.</dd>
      </dl><dl><dt>

          <code><a href="getDemographics.html">getDemographics()</a></code>

        </dt>
        <dd>Get demographic data</dd>
      </dl><dl><dt>

          <code><a href="getEmptyErrorLst.html">getEmptyErrorLst()</a></code>

        </dt>
        <dd>Creates a empty errorLst object</dd>
      </dl><dl><dt>

          <code><a href="getErrorTab.html">getErrorTab()</a></code>

        </dt>
        <dd>getErrorTab skeleton of list of errors</dd>
      </dl><dl><dt>

          <code><a href="getFocalAnimalPed.html">getFocalAnimalPed()</a></code>

        </dt>
        <dd>Get pedigree based on list of focal animals</dd>
      </dl><dl><dt>

          <code><a href="getGVGenotype.html">getGVGenotype()</a></code>

        </dt>
        <dd>Get Genetic Value Genotype data structure for reportGV function.</dd>
      </dl><dl><dt>

          <code><a href="getGVPopulation.html">getGVPopulation()</a></code>

        </dt>
        <dd>Get the population of interest for the Genetic Value analysis.</dd>
      </dl><dl><dt>

          <code><a href="getGenoDefinedParentGenotypes.html">getGenoDefinedParentGenotypes()</a></code>

        </dt>
        <dd>Assigns parental genotype contributions to an IDs genotype by attributing alleles to sire or dam</dd>
      </dl><dl><dt>

          <code><a href="getGenotypes.html">getGenotypes()</a></code>

        </dt>
        <dd>Get genotypes from file</dd>
      </dl><dl><dt>

          <code><a href="getIdsWithOneParent.html">getIdsWithOneParent()</a></code>

        </dt>
        <dd>getIdsWithOneParent extracts IDs of animals pedigree without either a sire or a dam</dd>
      </dl><dl><dt>

          <code><a href="getIncludeColumns.html">getIncludeColumns()</a></code>

        </dt>
        <dd>Get the superset of columns that can be in a pedigree file.</dd>
      </dl><dl><dt>

          <code><a href="getIndianOriginStatus.html">getIndianOriginStatus()</a></code>

        </dt>
        <dd>Get Indian-origin status of group</dd>
      </dl><dl><dt>

          <code><a href="getLkDirectAncestors.html">getLkDirectAncestors()</a></code>

        </dt>
        <dd>Get the direct ancestors of selected animals</dd>
      </dl><dl><dt>

          <code><a href="getLkDirectRelatives.html">getLkDirectRelatives()</a></code>

        </dt>
        <dd>Get the direct ancestors of selected animals</dd>
      </dl><dl><dt>

          <code><a href="getLogo.html">getLogo()</a></code>

        </dt>
        <dd>Get Logo file name</dd>
      </dl><dl><dt>

          <code><a href="getMaxAx.html">getMaxAx()</a></code>

        </dt>
        <dd>Get the maximum of the absolute values of the negative (males) and positive (female) animal counts.</dd>
      </dl><dl><dt>

          <code><a href="getMinParentAge.html">getMinParentAge()</a></code>

        </dt>
        <dd>Get minimum parent age.</dd>
      </dl><dl><dt>

          <code><a href="getOffspring.html">getOffspring()</a></code>

        </dt>
        <dd>Get offspring to corresponding animal IDs provided</dd>
      </dl><dl><dt>

          <code><a href="getParamDef.html">getParamDef()</a></code>

        </dt>
        <dd>Get parameter definitions from tokens found in configuration file.</dd>
      </dl><dl><dt>

          <code><a href="getParents.html">getParents()</a></code>

        </dt>
        <dd>Get parents to corresponding animal IDs provided</dd>
      </dl><dl><dt>

          <code><a href="getPedDirectRelatives.html">getPedDirectRelatives()</a></code>

        </dt>
        <dd>Get the direct ancestors of selected animals from supplied pedigree.</dd>
      </dl><dl><dt>

          <code><a href="getPedMaxAge.html">getPedMaxAge()</a></code>

        </dt>
        <dd>Get the maximum age of live animals in the pedigree.</dd>
      </dl><dl><dt>

          <code><a href="getPedigree.html">getPedigree()</a></code>

        </dt>
        <dd>Get pedigree from file</dd>
      </dl><dl><dt>

          <code><a href="getPossibleCols.html">getPossibleCols()</a></code>

        </dt>
        <dd>Get possible column names for a studbook.</dd>
      </dl><dl><dt>

          <code><a href="getPotentialParents.html">getPotentialParents()</a></code>

        </dt>
        <dd>Get the lists of portential parents for all individuals born in the colony with one or two unknown parents.</dd>
      </dl><dl><dt>

          <code><a href="getPotentialSires.html">getPotentialSires()</a></code>

        </dt>
        <dd>Provides list of potential sires</dd>
      </dl><dl><dt>

          <code><a href="getProbandPedigree.html">getProbandPedigree()</a></code>

        </dt>
        <dd>Gets pedigree to ancestors of provided group leaving uninformative ancestors.</dd>
      </dl><dl><dt>

          <code><a href="getProductionStatus.html">getProductionStatus()</a></code>

        </dt>
        <dd>Get production status of group</dd>
      </dl><dl><dt>

          <code><a href="getProportionLow.html">getProportionLow()</a></code>

        </dt>
        <dd>Get proportion of Low genetic value animals</dd>
      </dl><dl><dt>

          <code><a href="getPyramidAgeDist.html">getPyramidAgeDist()</a></code>

        </dt>
        <dd>Get the age distribution for the pedigree</dd>
      </dl><dl><dt>

          <code><a href="getPyramidPlot.html">getPyramidPlot()</a></code>

        </dt>
        <dd>Creates a pyramid plot of the pedigree provided.</dd>
      </dl><dl><dt>

          <code><a href="getRecordStatusIndex.html">getRecordStatusIndex()</a></code>

        </dt>
        <dd>Returns record numbers with selected <code>recordStatus</code>.</dd>
      </dl><dl><dt>

          <code><a href="getRequiredCols.html">getRequiredCols()</a></code>

        </dt>
        <dd>Get required column names for a studbook.</dd>
      </dl><dl><dt>

          <code><a href="getSexRatioWithAdditions.html">getSexRatioWithAdditions()</a></code>

        </dt>
        <dd>getSexRatioWithAdditions returns the sex ratio of a group.</dd>
      </dl><dl><dt>

          <code><a href="getSiteInfo.html">getSiteInfo()</a></code>

        </dt>
        <dd>Get site information</dd>
      </dl><dl><dt>

          <code><a href="getTokenList.html">getTokenList()</a></code>

        </dt>
        <dd>Gets tokens from character vector of lines</dd>
      </dl><dl><dt>

          <code><a href="getVersion.html">getVersion()</a></code>

        </dt>
        <dd>getVersion Get the version number of nprcgenekeepr</dd>
      </dl><dl><dt>

          <code><a href="get_and_or_list.html">get_and_or_list()</a></code>

        </dt>
        <dd>Returns a one element character string with correct punctuation for a list made up of the elements of the character vector argument.</dd>
      </dl><dl><dt>

          <code><a href="get_elapsed_time_str.html">get_elapsed_time_str()</a></code>

        </dt>
        <dd>Returns the elapsed time since start_time.</dd>
      </dl><dl><dt>

          <code><a href="groupAddAssign.html">groupAddAssign()</a></code>

        </dt>
        <dd>Add animals to an existing breeding group or forms groups:</dd>
      </dl><dl><dt>

          <code><a href="groupMembersReturn.html">groupMembersReturn()</a></code>

        </dt>
        <dd>Forms return list of groupAddAssign function</dd>
      </dl><dl><dt>

          <code><a href="hasBothParents.html">hasBothParents()</a></code>

        </dt>
        <dd>hasBothParents checks to see if both parents are identified.</dd>
      </dl><dl><dt>

          <code><a href="hasGenotype.html">hasGenotype()</a></code>

        </dt>
        <dd>Check for genotype data in dataframe</dd>
      </dl><dl><dt>

          <code><a href="headerDisplayNames.html">headerDisplayNames()</a></code>

        </dt>
        <dd>Convert internal column names to display or header names.</dd>
      </dl><dl><dt>

          <code><a href="initializeHaremGroups.html">initializeHaremGroups()</a></code>

        </dt>
        <dd>Make the initial groupMembers animal list</dd>
      </dl><dl><dt>

          <code><a href="insertChangedColsTab.html">insertChangedColsTab()</a></code>

        </dt>
        <dd>insertChangedColsTab insert a list of changed columns found by <code>qcStudbook</code> in the pedigree file</dd>
      </dl><dl><dt>

          <code><a href="insertErrorTab.html">insertErrorTab()</a></code>

        </dt>
        <dd>insertErrorTab insert a list of errors found by <code>qcStudbook</code> in the pedigree file</dd>
      </dl><dl><dt>

          <code><a href="insertSeparators.html">insertSeparators()</a></code>

        </dt>
        <dd>insertSeparators inserts the character "-" between year and month and between month and day portions of a date string in %Y%m%d format.</dd>
      </dl><dl><dt>

          <code><a href="isEmpty.html">isEmpty()</a></code>

        </dt>
        <dd>Is vector empty or all NA values.</dd>
      </dl><dl><dt>

          <code><a href="is_valid_date_str.html">is_valid_date_str()</a></code>

        </dt>
        <dd>Returns TRUE if the string is a valid date.</dd>
      </dl><dl><dt>

          <code><a href="kinMatrix2LongForm.html">kinMatrix2LongForm()</a></code>

        </dt>
        <dd>Reformats a kinship matrix into a long-format table.</dd>
      </dl><dl><dt>

          <code><a href="kinship.html">kinship()</a></code>

        </dt>
        <dd>Generates a kinship matrix.</dd>
      </dl><dl><dt>

          <code><a href="kinshipMatricesToKValues.html">kinshipMatricesToKValues()</a></code>

        </dt>
        <dd>Forms kValue matrix from list of kinship matrices</dd>
      </dl><dl><dt>

          <code><a href="kinshipMatrixToKValues.html">kinshipMatrixToKValues()</a></code>

        </dt>
        <dd>Extracts a dataframe with a row for each kinship coeficient in the kinship matrix</dd>
      </dl><dl><dt>

          <code><a href="lacy1989Ped.html">lacy1989Ped</a></code>

        </dt>
        <dd>lacy1989Ped small hypothetical pedigree</dd>
      </dl><dl><dt>

          <code><a href="lacy1989PedAlleles.html">lacy1989PedAlleles</a></code>

        </dt>
        <dd>lacy1989PedAlleles is a dataframe produced by <code>geneDrop</code> on <code>lacy1989Ped</code> with 5000 iterations.</dd>
      </dl><dl><dt>

          <code><a href="makeAvailable.html">makeAvailable()</a></code>

        </dt>
        <dd>Convenience function to make the initial available animal list</dd>
      </dl><dl><dt>

          <code><a href="makeCEPH.html">makeCEPH()</a></code>

        </dt>
        <dd>Make a CEPH-style pedigree for each id</dd>
      </dl><dl><dt>

          <code><a href="makeExamplePedigreeFile.html">makeExamplePedigreeFile()</a></code>

        </dt>
        <dd>Write copy of nprcgenekeepr::examplePedigree into a file</dd>
      </dl><dl><dt>

          <code><a href="makeGroupMembers.html">makeGroupMembers()</a></code>

        </dt>
        <dd>Convenience function to make the initial groupMembers animal list</dd>
      </dl><dl><dt>

          <code><a href="makeGrpNum.html">makeGrpNum()</a></code>

        </dt>
        <dd>Convenience function to make the initial grpNum list</dd>
      </dl><dl><dt>

          <code><a href="makeRelationClassesTable.html">makeRelationClassesTable()</a></code>

        </dt>
        <dd>Make relation classes table from <code>kin</code> dataframe.</dd>
      </dl><dl><dt>

          <code><a href="makeRoundUp.html">makeRoundUp()</a></code>

        </dt>
        <dd>Round up the provided integer vector <code>int</code> according to the <code>modulus</code>.</dd>
      </dl><dl><dt>

          <code><a href="makeSimPed.html">makeSimPed()</a></code>

        </dt>
        <dd>Makes a simulated pedigree using representative sires and dams</dd>
      </dl><dl><dt>

          <code><a href="makesLoop.html">makesLoop()</a></code>

        </dt>
        <dd><code>makesLoop</code> tests for a common ancestor.</dd>
      </dl><dl><dt>

          <code><a href="mapIdsToObfuscated.html">mapIdsToObfuscated()</a></code>

        </dt>
        <dd>Map IDs to Obfuscated IDs</dd>
      </dl><dl><dt>

          <code><a href="meanKinship.html">meanKinship()</a></code>

        </dt>
        <dd>Calculates the mean kinship for each animal in a kinship matrix</dd>
      </dl><dl><dt>

          <code><a href="nprcgenekeepr.html">nprcgenekeepr</a></code>

        </dt>
        <dd>Genetic Tools for Colony Management</dd>
      </dl><dl><dt>

          <code><a href="obfuscateDate.html">obfuscateDate()</a></code>

        </dt>
        <dd>obfucateDate adds a random number of days bounded by plus and minus max delta</dd>
      </dl><dl><dt>

          <code><a href="obfuscateId.html">obfuscateId()</a></code>

        </dt>
        <dd>obfucateId creates a vector of ID aliases of specified length</dd>
      </dl><dl><dt>

          <code><a href="obfuscatePed.html">obfuscatePed()</a></code>

        </dt>
        <dd>obfuscatePed takes a pedigree object and creates aliases for all IDs and adjusts all date within a specified amount.</dd>
      </dl><dl><dt>

          <code><a href="offspringCounts.html">offspringCounts()</a></code>

        </dt>
        <dd>Finds the total number of offspring for each animal in the pedigree</dd>
      </dl><dl><dt>

          <code><a href="orderReport.html">orderReport()</a></code>

        </dt>
        <dd>Order the results of the genetic value analysis for use in a report.</dd>
      </dl><dl><dt>

          <code><a href="ped1Alleles.html">ped1Alleles</a></code>

        </dt>
        <dd>ped1Alleles is a dataframe created by the geneDrop function</dd>
      </dl><dl><dt>

          <code><a href="pedDuplicateIds.html">pedDuplicateIds</a></code>

        </dt>
        <dd>pedDuplicateIds is a dataframe with 9 rows and 5 columns (ego_id, sire, dam_id, sex, birth_date) representing a full pedigree with a duplicated record.</dd>
      </dl><dl><dt>

          <code><a href="pedFemaleSireMaleDam.html">pedFemaleSireMaleDam</a></code>

        </dt>
        <dd>pedFemaleSireMaleDam is a dataframe with 8 rows and 5 columns (ego_id, sire, dam_id, sex, birth_date) representing a full pedigree with the errors of having a sire labeled as female and a dam labeled as male.</dd>
      </dl><dl><dt>

          <code><a href="pedGood.html">pedGood</a></code>

        </dt>
        <dd>pedGood is a dataframe with 8 rows and 5 columns (ego_id, sire, dam_id, sex, birth_date) representing a full pedigree with no errors.</dd>
      </dl><dl><dt>

          <code><a href="pedInvalidDates.html">pedInvalidDates</a></code>

        </dt>
        <dd>pedInvalidDates is a dataframe with 8 rows and 5 columns (ego_id, sire, dam_id, sex, birth_date) representing a full pedigree with values in the <code>birth_date</code> column that are not valid dates.</dd>
      </dl><dl><dt>

          <code><a href="pedMissingBirth.html">pedMissingBirth</a></code>

        </dt>
        <dd>pedMissingBirth is a dataframe with 8 rows and 5 columns (ego_id, sire, dam_id, sex, birth_date) representing a full pedigree with no errors.</dd>
      </dl><dl><dt>

          <code><a href="pedOne.html">pedOne</a></code>

        </dt>
        <dd>pedOne is a loadable version of a pedigree file fragment used for testing and demonstration</dd>
      </dl><dl><dt>

          <code><a href="pedSameMaleIsSireAndDam.html">pedSameMaleIsSireAndDam</a></code>

        </dt>
        <dd>pedSameMaleIsSireAndDam is a dataframe with 8 rows and 5 columns (ego_id, sire, dam_id, sex, birth_date) representing a full pedigree with no errors.</dd>
      </dl><dl><dt>

          <code><a href="pedSix.html">pedSix</a></code>

        </dt>
        <dd>pedSix is a loadable version of a pedigree file fragment used for testing and demonstration</dd>
      </dl><dl><dt>

          <code><a href="pedWithGenotype.html">pedWithGenotype</a></code>

        </dt>
        <dd>pedWithGenotype is a dataframe produced from qcPed by adding made up genotypes.</dd>
      </dl><dl><dt>

          <code><a href="pedWithGenotypeReport.html">pedWithGenotypeReport</a></code>

        </dt>
        <dd>pedWithGenotypeReport is a list containing the output of <code>reportGV</code>.</dd>
      </dl><dl><dt>

          <code><a href="print.html">print(<i>&lt;summary.nprcgenekeeprErr&gt;</i>)</a></code> <code><a href="print.html">print(<i>&lt;summary.nprcgenekeeprGV&gt;</i>)</a></code>

        </dt>
        <dd>print.summary.nprcgenekeepr print.summary.nprcgenekeeprGV</dd>
      </dl><dl><dt>

          <code><a href="qcBreeders.html">qcBreeders</a></code>

        </dt>
        <dd>qcBreeders is a list of 29 baboon IDs that are potential breeders</dd>
      </dl><dl><dt>

          <code><a href="qcPed.html">qcPed</a></code>

        </dt>
        <dd>qcPed is a dataframe with 277 rows and 6 columns</dd>
      </dl><dl><dt>

          <code><a href="qcPedGvReport.html">qcPedGvReport</a></code>

        </dt>
        <dd>qcPedGvReport is a genetic value report</dd>
      </dl><dl><dt>

          <code><a href="qcStudbook.html">qcStudbook()</a></code>

        </dt>
        <dd>Quality Control for the Studbook or pedigree</dd>
      </dl><dl><dt>

          <code><a href="rankSubjects.html">rankSubjects()</a></code>

        </dt>
        <dd>Ranks animals based on genetic value.</dd>
      </dl><dl><dt>

          <code><a href="rbindFill.html">rbindFill()</a></code>

        </dt>
        <dd>Append the rows of one dataframe to another.</dd>
      </dl><dl><dt>

          <code><a href="readExcelPOSIXToCharacter.html">readExcelPOSIXToCharacter()</a></code>

        </dt>
        <dd>Read in Excel file and convert POSIX dates to character</dd>
      </dl><dl><dt>

          <code><a href="removeDuplicates.html">removeDuplicates()</a></code>

        </dt>
        <dd>Remove duplicate records from pedigree</dd>
      </dl><dl><dt>

          <code><a href="removeEarlyDates.html">removeEarlyDates()</a></code>

        </dt>
        <dd>removeEarlyDates removes dates before a specified year</dd>
      </dl><dl><dt>

          <code><a href="removeGroupIfNoAvailableAnimals.html">removeGroupIfNoAvailableAnimals()</a></code>

        </dt>
        <dd>Remove group numbers when all available animals have been used</dd>
      </dl><dl><dt>

          <code><a href="removePotentialSires.html">removePotentialSires()</a></code>

        </dt>
        <dd>Removes potential sires from list of Ids</dd>
      </dl><dl><dt>

          <code><a href="removeSelectedAnimalFromAvailableAnimals.html">removeSelectedAnimalFromAvailableAnimals()</a></code>

        </dt>
        <dd>Updates list of available animals by removing the selected animal</dd>
      </dl><dl><dt>

          <code><a href="removeUninformativeFounders.html">removeUninformativeFounders()</a></code>

        </dt>
        <dd>Remove uninformative founders.</dd>
      </dl><dl><dt>

          <code><a href="removeUnknownAnimals.html">removeUnknownAnimals()</a></code>

        </dt>
        <dd>removeUnknownAnimals Removes unknown animals added to pedigree that serve as placeholders for unknown parents.</dd>
      </dl><dl><dt>

          <code><a href="remove_auto_gen_ids.html">remove_auto_gen_ids()</a></code>

        </dt>
        <dd>Remove automatically generated IDs from pedigree</dd>
      </dl><dl><dt>

          <code><a href="reportGV.html">reportGV()</a></code>

        </dt>
        <dd>Generates a genetic value report for a provided pedigree.</dd>
      </dl><dl><dt>

          <code><a href="resetGroup.html">resetGroup()</a></code>

        </dt>
        <dd>Update or add the "group" field of a Pedigree.</dd>
      </dl><dl><dt>

          <code><a href="rhesusGenotypes.html">rhesusGenotypes</a></code>

        </dt>
        <dd>rhesusGenotypes is a dataframe with two haplotypes per animal</dd>
      </dl><dl><dt>

          <code><a href="rhesusPedigree.html">rhesusPedigree</a></code>

        </dt>
        <dd>rhesusPedigree is a pedigree object</dd>
      </dl><dl><dt>

          <code><a href="runGeneKeepR.html">runGeneKeepR()</a></code>

        </dt>
        <dd>Allows running <code>shiny</code> application with <code><a href="../reference/runGeneKeepR.html">nprcgenekeepr::runGeneKeepR()</a></code></dd>
      </dl><dl><dt>

          <code><a href="saveDataframesAsFiles.html">saveDataframesAsFiles()</a></code>

        </dt>
        <dd>Write copy of dataframes to either CSV, TXT, or Excel file.</dd>
      </dl><dl><dt>

          <code><a href="setExit.html">setExit()</a></code>

        </dt>
        <dd>Sets the exit date, if there is no exit column in the table</dd>
      </dl><dl><dt>

          <code><a href="setPopulation.html">setPopulation()</a></code>

        </dt>
        <dd>Population designation function</dd>
      </dl><dl><dt>

          <code><a href="set_seed.html">set_seed()</a></code>

        </dt>
        <dd>Work around for unit tests using sample() among various versions of R</dd>
      </dl><dl><dt>

          <code><a href="smallPed.html">smallPed</a></code>

        </dt>
        <dd>smallPed is a hypothetical pedigree</dd>
      </dl><dl><dt>

          <code><a href="smallPedTree.html">smallPedTree</a></code>

        </dt>
        <dd>smallPedTree is a pedigree tree made from <code>smallPed</code></dd>
      </dl><dl><dt>

          <code><a href="str_detect_fixed_all.html">str_detect_fixed_all()</a></code>

        </dt>
        <dd>Returns a logical vector with results of stri_detect() for each pattern in second parameters character vector.</dd>
      </dl><dl><dt>

          <code><a href="summarizeKinshipValues.html">summarizeKinshipValues()</a></code>

        </dt>
        <dd>Summary statistics for imputed kinship values</dd>
      </dl><dl><dt>

          <code><a href="summary.html">summary(<i>&lt;nprcgenekeeprErr&gt;</i>)</a></code> <code><a href="summary.html">summary(<i>&lt;nprcgenekeeprGV&gt;</i>)</a></code>

        </dt>
        <dd>summary.nprcgenekeeprErr Summary function for class nprcgenekeeprErr</dd>
      </dl><dl><dt>

          <code><a href="toCharacter.html">toCharacter()</a></code>

        </dt>
        <dd>Force dataframe columns to character</dd>
      </dl><dl><dt>

          <code><a href="trimPedigree.html">trimPedigree()</a></code>

        </dt>
        <dd>Trim pedigree to ancestors of provided group by removing uninformative individuals</dd>
      </dl><dl><dt>

          <code><a href="unknown2NA.html">unknown2NA()</a></code>

        </dt>
        <dd>Removing IDs having "UNKNOWN" regardless of case</dd>
      </dl><dl><dt>

          <code><a href="withinIntegerRange.html">withinIntegerRange()</a></code>

        </dt>
        <dd>Get integer within a range</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Michael Raboin, Terry Therneau, Amanda Vinson, R. Mark Sharp, Matthew Schultz, Southwest National Primate Research Center NIH grant P51 RR13986, Oregon National Primate Research Center grant P51 OD011092.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

