<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Genetic Management Functions — nprcgenekeepr-package • nprcgenekeepr</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Genetic Management Functions — nprcgenekeepr-package"><meta name="description" content="Primary Data Structure — Pedigree
Contains studbook information for a number of individuals.
ASSUME: All IDs listed in the sire or dam columns must have a row entry in
the id column

Provides genetic tools for colony management and is a derivation of the work in Amanda Vinson and Michael J Raboin (2015) https://pmc.ncbi.nlm.nih.gov/articles/PMC4671785/ &quot;A Practical Approach for Designing Breeding Groups to Maximize Genetic Diversity in a Large Colony of Captive Rhesus Macaques ('Macaca' 'mulatto')&quot;. It provides a 'Shiny' application with an exposed API. The application supports five groups of functions: (1) Quality control of studbooks contained in text files or 'Excel' workbooks and of pedigrees within 'LabKey' Electronic Health Records (EHR); (2) Creation of pedigrees from a list of animals using the 'LabKey' EHR integration; (3) Creation and display of an age by sex pyramid plot of the living animals within the designated pedigree; (4) Generation of genetic value analysis reports; and (5) Creation of potential breeding groups with and without proscribed sex ratios and defined maximum kinships."><meta property="og:description" content="Primary Data Structure — Pedigree
Contains studbook information for a number of individuals.
ASSUME: All IDs listed in the sire or dam columns must have a row entry in
the id column

Provides genetic tools for colony management and is a derivation of the work in Amanda Vinson and Michael J Raboin (2015) https://pmc.ncbi.nlm.nih.gov/articles/PMC4671785/ &quot;A Practical Approach for Designing Breeding Groups to Maximize Genetic Diversity in a Large Colony of Captive Rhesus Macaques ('Macaca' 'mulatto')&quot;. It provides a 'Shiny' application with an exposed API. The application supports five groups of functions: (1) Quality control of studbooks contained in text files or 'Excel' workbooks and of pedigrees within 'LabKey' Electronic Health Records (EHR); (2) Creation of pedigrees from a list of animals using the 'LabKey' EHR integration; (3) Creation and display of an age by sex pyramid plot of the living animals within the designated pedigree; (4) Generation of genetic value analysis reports; and (5) Creation of potential breeding groups with and without proscribed sex ratios and defined maximum kinships."><meta property="og:image" content="https://github.com/rmsharp/nprcgenekeepr/reference/figures/card.png"><meta property="og:image:alt" content="Conserve Genes"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">nprcgenekeepr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.7</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/a2interactive.html">Interactive Use of nprcgenekeepr</a></li>
    <li><a class="dropdown-item" href="../articles/a3manual.html">Genetic Management Tools Manual</a></li>
    <li><a class="dropdown-item" href="../articles/ColonyManagerTutorial.html">GeneKeepR: A Colony Manager's Tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/simulatedKValues.html">Quick Example of Simulated Kinships with Partial Parentage</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/rmsharp/nprcgenekeepr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Genetic Management Functions</h1>
      <small class="dont-index">Source: <a href="https://github.com/rmsharp/nprcgenekeepr/blob/v1.0.7/R/nprcgenekeeper.R"><code>R/nprcgenekeeper.R</code></a>, <a href="https://github.com/rmsharp/nprcgenekeepr/blob/v1.0.7/R/nprcgenekeepr-package.R"><code>R/nprcgenekeepr-package.R</code></a></small>
      <div class="d-none name"><code>nprcgenekeepr-package.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Primary Data Structure — Pedigree</p>
<p>Contains studbook information for a number of individuals.
ASSUME: All IDs listed in the sire or dam columns must have a row entry in
the id column</p>
<p></p>
<p>Provides genetic tools for colony management and is a derivation of the work in Amanda Vinson and Michael J Raboin (2015) <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC4671785/" class="external-link">https://pmc.ncbi.nlm.nih.gov/articles/PMC4671785/</a> "A Practical Approach for Designing Breeding Groups to Maximize Genetic Diversity in a Large Colony of Captive Rhesus Macaques ('Macaca' 'mulatto')". It provides a 'Shiny' application with an exposed API. The application supports five groups of functions: (1) Quality control of studbooks contained in text files or 'Excel' workbooks and of pedigrees within 'LabKey' Electronic Health Records (EHR); (2) Creation of pedigrees from a list of animals using the 'LabKey' EHR integration; (3) Creation and display of an age by sex pyramid plot of the living animals within the designated pedigree; (4) Generation of genetic value analysis reports; and (5) Creation of potential breeding groups with and without proscribed sex ratios and defined maximum kinships.</p>
    </div>


    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="getIncludeColumns.html">getIncludeColumns</a></code> to get set of columns that can
be used in a pedigree file</p>
<p>A Pedigree is a data frame within the <code>R</code> environment with the following
possible columns:</p><ul><li><p>{id} {– character vector with unique identifier for an individual}</p></li>
<li><p>{sire} {– character vector with unique identifier for an
individual's father (<code>NA</code> if unknown).}</p></li>
<li><p>{dam} {– character vector with unique identifier for an
individual's mother (<code>NA</code> if unknown).}</p></li>
<li><p>{sex} {– factor {levels: "M", "F", "U"} Sex specifier for an
individual}</p></li>
<li><p>{gen} {– integer vector with the generation number of the
individual}</p></li>
<li><p>{birth} {– Date or <code>NA</code> (optional) with the individual's birth
date}</p></li>
<li><p>{exit} {– Date or <code>NA</code> (optional) with the individual's exit
date (death, or departure if applicable)}</p></li>
<li><p>{ancestry} {– character vector or <code>NA</code> (optional) that
indicates the geographic population to which the individual belongs.}</p></li>
<li><p>{age} {– numeric or <code>NA</code> (optional) indicating the individual's
current age or age at exit.}</p></li>
<li><p>{population} {– logical (optional)
Is the id part of the extant population?}</p></li>
<li><p>{origin} {– character vector or <code>NA</code> (optional) that indicates
the name of the facility that the individual was imported from if other than
local.<code>NA</code> indicates the individual was not imported.}</p></li>
</ul><p>Pedigree File Testing Functions</p>
<ul><li><p>{<a href="qcStudbook.html">qcStudbook</a>} {— Main pedigree curation function that
performs basic quality control on pedigree information}</p></li>
<li><p>{<a href="fixColumnNames.html">fixColumnNames</a>} {— Changes original column names and into
standardized names.}</p></li>
<li><p>{<a href="checkRequiredCols.html">checkRequiredCols</a>} {— Examines column names, cols, to see
if all required column names are present.}</p></li>
<li><p>{<a href="correctParentSex.html">correctParentSex</a>} {— Sets sex for animals listed as either
a sire or dam.}</p></li>
<li><p>{<a href="getDateErrorsAndConvertDatesInPed.html">getDateErrorsAndConvertDatesInPed</a>} {— Converts columns of
dates in text form to <code>Date</code> object columns}</p></li>
<li><p>{<a href="checkParentAge.html">checkParentAge</a>} {— Check parent ages to be at least
<code>minParentAge</code>}</p></li>
<li><p>{<a href="removeDuplicates.html">removeDuplicates</a>} {— Remove duplicate records from
pedigree}</p></li>
</ul><p>Gene Dropping Function</p>
<ul><li><p>{<a href="geneDrop.html">geneDrop</a>} {— Performs a gene drop simulation based on the
provided pedigree information}</p></li>
</ul><p>Genetic Value Analysis Functions</p>
<p>Contains functions to calculate the kinship coefficient and genome
uniqueness for animals listed in a Pedigree table.</p><ul><li><p>{<a href="meanKinship.html">meanKinship</a>} {— Calculates the mean kinship for each animal
in a kinship matrix}</p></li>
<li><p>{<a href="calcA.html">calcA</a>} {— Calculates <code>a</code>, the number of an
individual's alleles that are rare in each simulation.}</p></li>
<li><p>{<a href="alleleFreq.html">alleleFreq</a>} {— Calculates the count of each allele in the
provided vector.}</p></li>
<li><p>{<a href="calcFE.html">calcFE</a>} {— Calculates founder equivalents.}</p></li>
<li><p>{<a href="calcFG.html">calcFG</a>} {— Calculates founder genome equivalents.}</p></li>
<li><p>{<a href="calcFEFG.html">calcFEFG</a>} {— Returns founder equivalents <code>FE</code> and
<code>FG</code> as elements in a list.}</p></li>
<li><p>{<a href="calcGU.html">calcGU</a>} {— Calculates genome uniqueness for each ID that is
part of the population.}</p></li>
<li><p>{<a href="geneDrop.html">geneDrop</a>} {— Performs a gene drop simulation based on the
pedigree information.}</p></li>
<li><p>{<a href="chooseAlleles.html">chooseAlleles</a>} {— Combines two vectors of alleles by
randomly selecting one allele or the other at each position.}</p></li>
<li><p>{<a href="calcRetention.html">calcRetention</a>} {— Calculates allelic retention.}</p></li>
<li><p>{<a href="filterKinMatrix.html">filterKinMatrix</a>} {— Filters a kinship matrix to include
only the egos listed in 'ids'}</p></li>
<li><p>{<a href="kinship.html">kinship</a>} {— Generates a kinship matrix}</p></li>
<li><p>{<a href="reportGV.html">reportGV</a>} {— Generates a genetic value report for a
provided pedigree.}</p></li>
</ul><p>Plotting Functions</p>
<ul><li><p>{<a href="meanKinship.html">meanKinship</a>} {— Calculates the mean kinship for each animal
in a kinship matrix}</p></li>
</ul><p>Breeding Group Formation Functions</p>
<ul><li><p>{<a href="meanKinship.html">meanKinship</a>} {— Calculates the mean kinship for each animal
in a kinship matrix}</p></li>
</ul><p>Useful links:</p><ul><li><p><a href="https://rmsharp.github.io/nprcgenekeepr/" class="external-link">https://rmsharp.github.io/nprcgenekeepr/</a></p></li>
<li><p><a href="https://github.com/rmsharp/nprcgenekeepr">https://github.com/rmsharp/nprcgenekeepr</a></p></li>
<li><p>Report bugs at <a href="https://github.com/rmsharp/nprcgenekeepr/issues">https://github.com/rmsharp/nprcgenekeepr/issues</a></p></li>
</ul></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p><strong>Maintainer</strong>: R. Mark Sharp <a href="mailto:rmsharp@me.com">rmsharp@me.com</a> (<a href="https://orcid.org/0000-0002-6170-6942" class="external-link">ORCID</a>) [copyright holder, data contributor]</p>
<p>Authors:</p><ul><li><p>Michael Raboin</p></li>
<li><p>Terry Therneau</p></li>
<li><p>Amanda Vinson [data contributor]</p></li>
<li><p>Matthew Schultz (<a href="https://orcid.org/0000-0001-5103-4305" class="external-link">ORCID</a>)</p></li>
</ul><p>Other contributors:</p><ul><li><p>Southwest National Primate Research Center NIH grant P51 RR13986 [funder]</p></li>
<li><p>Oregon National Primate Research Center grant P51 OD011092 [funder]</p></li>
</ul></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Michael Raboin, Terry Therneau, Amanda Vinson, R. Mark Sharp, Matthew Schultz, Southwest National Primate Research Center NIH grant P51 RR13986, Oregon National Primate Research Center grant P51 OD011092.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

