<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Gene drop simulation based on the provided pedigree information — geneDrop • nprcgenekeepr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Gene drop simulation based on the provided pedigree information — geneDrop" />
<meta property="og:description" content="Part of Genetic Value Analysis" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">nprcgenekeepr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/ColonyManagerTutorial.html">GeneKeepR: A Colony Manager's Tutorial</a>
    </li>
    <li>
      <a href="../articles/a2interactive.html">Interactive Use of nprcgenekeepr</a>
    </li>
    <li>
      <a href="../articles/a3manual.html">Genetic Management Tools Manual</a>
    </li>
    <li>
      <a href="../articles/shiny_app_use.html">GeneKeepR Tutorial</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/rmsharp/nprcgenekeepr/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Gene drop simulation based on the provided pedigree information</h1>
    <small class="dont-index">Source: <a href='https://github.com/rmsharp/nprcgenekeepr/blob/master/R/geneDrop.R'><code>R/geneDrop.R</code></a></small>
    <div class="hidden name"><code>geneDrop.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Part of Genetic Value Analysis</p>
    </div>

    <pre class="usage"><span class='fu'>geneDrop</span><span class='op'>(</span>
  <span class='va'>ids</span>,
  <span class='va'>sires</span>,
  <span class='va'>dams</span>,
  <span class='va'>gen</span>,
  genotype <span class='op'>=</span> <span class='cn'>NULL</span>,
  n <span class='op'>=</span> <span class='fl'>5000</span>,
  updateProgress <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>ids</th>
      <td><p>A character vector of IDs for a set of animals.</p></td>
    </tr>
    <tr>
      <th>sires</th>
      <td><p>A character vector with IDS of the sires for the set of
animals. <code>NA</code> is used for missing sires.</p></td>
    </tr>
    <tr>
      <th>dams</th>
      <td><p>A character vector with IDS of the dams for the set of
animals. <code>NA</code> is used for missing dams.</p></td>
    </tr>
    <tr>
      <th>gen</th>
      <td><p>An integer vector indicating the generation number for each
animal.</p></td>
    </tr>
    <tr>
      <th>genotype</th>
      <td><p>A dataframe containing known genotypes. It has three
columns:  <code>id</code>, <code>first</code>, and <code>second</code>. The second and third
columns contain the integers indicating the observed genotypes.</p>
<p>The gene dropping method from <em>Pedigree analysis by computer simulation</em>
by Jean W MacCluer, John L Vandeberg, and Oliver A Ryder (1986)
&lt;doi:10.1002/zoo.1430050209&gt; is used in the genetic value calculations.</p>
<p>Currently there is no means of handling knowing only one haplotype.
It will be easy to add another column to handle situations where only one
allele is observed and it is not known to be homozygous or heterozygous. The
new fourth column could have a frequency for homozygosity that could be
used in the gene dropping algorithm.</p>
<p>The genotypes are using indirection (integer instead of character) to
indicate the genes because the manipulation of character strings was found
to take 20-35 times longer to perform.</p>
<p>Adding additional columns to <code>genotype</code> does not significantly affect
the time require. Thus, it is convenient to add the corresponding haplotype
names to the dataframe using <code>first_name</code> and <code>second_name</code>.</p></td>
    </tr>
    <tr>
      <th>n</th>
      <td><p>integer indicating the number of iterations to simulate.
Default is 5000.</p></td>
    </tr>
    <tr>
      <th>updateProgress</th>
      <td><p>function or NULL. If this function is defined, it
will be called during each iteration to update a
<code><a href='https://rdrr.io/pkg/shiny/man/Progress.html'>shiny::Progress</a></code> object.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A data.frame <code>id, parent, V1 ... Vn</code>
A data.frame providing the maternal and paternal alleles for an animal
for each iteration. The first two columns provide the animal's ID and
whether the allele came from the sire or dam. These are followed by
<code>n</code> columns indicating the allele for that iteration.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># \donttest{</span>
<span class='co'>## We usually defined `n` to be &gt;= 5000</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://rmsharp.github.io/nprcgenekeepr/'>nprcgenekeepr</a></span><span class='op'>)</span>
<span class='va'>ped</span> <span class='op'>&lt;-</span> <span class='fu'>nprcgenekeepr</span><span class='fu'>::</span><span class='va'><a href='lacy1989Ped.html'>lacy1989Ped</a></span>
<span class='va'>allelesNew</span> <span class='op'>&lt;-</span> <span class='fu'>geneDrop</span><span class='op'>(</span><span class='va'>ped</span><span class='op'>$</span><span class='va'>id</span>, <span class='va'>ped</span><span class='op'>$</span><span class='va'>sire</span>, <span class='va'>ped</span><span class='op'>$</span><span class='va'>dam</span>, <span class='va'>ped</span><span class='op'>$</span><span class='va'>gen</span>,
                      genotype <span class='op'>=</span> <span class='cn'>NULL</span>, n <span class='op'>=</span> <span class='fl'>50</span>, updateProgress <span class='op'>=</span> <span class='cn'>NULL</span><span class='op'>)</span>
<span class='va'>genotype</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>id <span class='op'>=</span> <span class='va'>ped</span><span class='op'>$</span><span class='va'>id</span>,
                       first_allele <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='cn'>NA</span>, <span class='cn'>NA</span>, <span class='st'>"A001_B001"</span>, <span class='st'>"A001_B002"</span>,
                                        <span class='cn'>NA</span>, <span class='st'>"A001_B002"</span>, <span class='st'>"A001_B001"</span><span class='op'>)</span>,
                       second_allele <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='cn'>NA</span>, <span class='cn'>NA</span>, <span class='st'>"A010_B001"</span>, <span class='st'>"A001_B001"</span>,
                                         <span class='cn'>NA</span>, <span class='cn'>NA</span>, <span class='cn'>NA</span><span class='op'>)</span>,
                       stringsAsFactors <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
<span class='va'>pedWithGenotype</span> <span class='op'>&lt;-</span> <span class='fu'><a href='addGenotype.html'>addGenotype</a></span><span class='op'>(</span><span class='va'>ped</span>, <span class='va'>genotype</span><span class='op'>)</span>
<span class='va'>pedGenotype</span> <span class='op'>&lt;-</span> <span class='fu'><a href='getGVGenotype.html'>getGVGenotype</a></span><span class='op'>(</span><span class='va'>pedWithGenotype</span><span class='op'>)</span>
<span class='va'>allelesNewGen</span> <span class='op'>&lt;-</span> <span class='fu'>geneDrop</span><span class='op'>(</span><span class='va'>ped</span><span class='op'>$</span><span class='va'>id</span>, <span class='va'>ped</span><span class='op'>$</span><span class='va'>sire</span>, <span class='va'>ped</span><span class='op'>$</span><span class='va'>dam</span>, <span class='va'>ped</span><span class='op'>$</span><span class='va'>gen</span>,
                         genotype <span class='op'>=</span> <span class='va'>pedGenotype</span>,
                         n <span class='op'>=</span> <span class='fl'>5</span>, updateProgress <span class='op'>=</span> <span class='cn'>NULL</span><span class='op'>)</span>
<span class='co'># }</span>

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Michael Raboin, Terry Therneau, Amanda Vinson, R. Mark Sharp, Southwest National Primate Research Center NIH grant P51 RR13986, Oregon National Primate Research Center grant P51 OD011092.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


