<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Interactive Use of nprcgenekeepr • nprcgenekeepr</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Interactive Use of nprcgenekeepr">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">nprcgenekeepr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.7</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/a2interactive.html">Interactive Use of nprcgenekeepr</a></li>
    <li><a class="dropdown-item" href="../articles/a3manual.html">Genetic Management Tools Manual</a></li>
    <li><a class="dropdown-item" href="../articles/ColonyManagerTutorial.html">GeneKeepR: A Colony Manager's Tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/simulatedKValues.html">Quick Example of Simulated Kinships with Partial Parentage</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/rmsharp/nprcgenekeepr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Interactive Use of nprcgenekeepr</h1>
            <h3 data-toc-skip class="subtitle">An R Package for the
Genetic Management of Colonies</h3>
            
      
      <small class="dont-index">Source: <a href="https://github.com/rmsharp/nprcgenekeepr/blob/master/vignettes/a2interactive.Rmd"><code>vignettes/a2interactive.Rmd</code></a></small>
      <div class="d-none name"><code>a2interactive.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This tutorial demonstrates the major functions used within the Shiny
application provided by the <strong>nprcgenekeepr</strong> package and
provides sufficient insight into those functions that they may be used
independently.</p>
<p>This tutorial is primarily directed toward someone with experience
using R who wants to better understand how the Shiny application works
or to perform some actions not directly supported by the Shiny
application.</p>
<p>Please provide any comments, questions, or bug reports through the
GitHub issue tracker at <a href="https://github.com/rmsharp/nprcgenekeepr/issues">https://github.com/rmsharp/nprcgenekeepr/issues</a>.</p>
</div>
<div class="section level2">
<h2 id="installation-and-help">Installation and Help<a class="anchor" aria-label="anchor" href="#installation-and-help"></a>
</h2>
<p>You can install <strong>nprcgenekeepr</strong> from GitHub with the
following code.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="va">nprcgenekeepr</span><span class="op">)</span></span>
<span><span class="co">## Use the following code to get the development version</span></span>
<span><span class="co"># install.packages("devtools")</span></span>
<span><span class="co"># devtools::install_github("rmsharp/nprcgenekeepr")</span></span></code></pre></div>
<p>All missing dependencies should be automatically installed.</p>
<p>You can get help from the <strong>R</strong> console with</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="va">nprcgenekeepr</span></span></code></pre></div>
<p><strong>The help provided by this (<em>nprcgenekeepr.R</em>) needs to
be more complete and include links to the tutorials.</strong></p>
</div>
<div class="section level2">
<h2 id="reading-in-a-pedigree">Reading in a Pedigree<a class="anchor" aria-label="anchor" href="#reading-in-a-pedigree"></a>
</h2>
<p>A pedigrees can be imported using either Excel worksheets or text
files that contain all of the pedigree information or using either Excel
worksheets or text files that contain a list of focal animals with the
remainder of the pedigree information is pulled in through the LabKey
API.</p>
<p>This tutorial will use a pedigree file that can be created using the
<strong>makeExamplePedigreeFile</strong> function as shown below. The
function <strong>makeExamplePedigreeFile</strong> both saves a file and
returns the full path name to the saved file, which we are saving into
the variable <em>pedigreeFile</em>. Note: the user will select where to
store the file.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rmsharp.github.io/nprcgenekeepr/" class="external-link">nprcgenekeepr</a></span><span class="op">)</span></span>
<span><span class="va">pedigreeFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeExamplePedigreeFile.html">makeExamplePedigreeFile</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>This writes <em>ExamplePedigree.csv</em> to a place you select within
your file system.</p>
<p>You use the file name provided by the
<strong>makeExamplePedigreeFile</strong> function to tell
<strong>read.table</strong> what file to read.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">breederPedCsv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">pedigreeFile</span>,</span>
<span>  sep <span class="op">=</span> <span class="st">","</span>, header <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">breederPedCsv</span><span class="op">$</span><span class="va">fromCenter</span> <span class="op">&lt;-</span> <span class="st">"TRUE"</span></span>
<span><span class="va">breederPedCsv</span><span class="op">$</span><span class="va">fromCenter</span><span class="op">[</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">breederPedCsv</span><span class="op">$</span><span class="va">sire</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">breederPedCsv</span><span class="op">$</span><span class="va">dam</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fl">0.8</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">breederPedCsv</span><span class="op">$</span><span class="va">sire</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">breederPedCsv</span><span class="op">$</span><span class="va">dam</span><span class="op">)</span></span>
<span>    <span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span><span class="op">)</span></span>
<span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"FALSE"</span></span></code></pre></div>
<p>Note the number of rows read. Each row represents an individual
within the pedigree.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">breederPedCsv</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 3694</span></span></code></pre>
<p>The next step is to put the information read from the file into a
pedigree object. This is done with the <strong>qcStudbook</strong>
function, which examines the file contents and tests for common pedigree
errors.</p>
<p>You can see the errors that can be detected by
<strong>qcStudbook</strong> by returning the empty error list with
<strong>getEmptyErrorLst()</strong>. We are not showing the output of
the function call now because later in this tutorial we will explore
errors in more depth.</p>
<p><strong>qcStudbook</strong> can take four arguments <em>sb</em>,
<em>minParentAge</em> (in years), <em>reportChanges</em>, and
<em>reportErrors</em>. However, all but <em>sb</em> have default values
and only the <em>sb</em> argument is required.</p>
<p>It is prudent to ensure that parents are at least of breeding age,
which is species specific. I have used a <em>minParentAge</em> of 2
years.<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;Setting the &lt;em&gt;minParentAge&lt;/em&gt; to 3.5 and above will
cause an error along with the creation of a file
&lt;em&gt;~/lowParentAge.csv&lt;/em&gt; that will list the parents with low age at
the birth of an offspring.&lt;/p&gt;"><sup>1</sup></a></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">breederPed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qcStudbook.html">qcStudbook</a></span><span class="op">(</span><span class="va">breederPedCsv</span>, minParentAge <span class="op">=</span> <span class="fl">2L</span><span class="op">)</span></span></code></pre></div>
<p>If <strong>qcStudbook</strong> reports an error, change the call by
adding the <strong>reportErrors</strong> argument set to
<strong>TRUE</strong> and examine the returned object. More on this is
presented in the <strong>Pedigree Errors</strong> section.</p>
</div>
<div class="section level2">
<h2 id="identifying-focal-animals">Identifying Focal Animals<a class="anchor" aria-label="anchor" href="#identifying-focal-animals"></a>
</h2>
<p>You may want to focus your work on a <em>focal</em> group of animals.
This can be done by reading in a list of animal IDs that make up the
<em>focal</em> group and use that list to update the pedigree.
Alternatively you can created a list of animal IDs based on criteria you
have selected.</p>
<p>For example, to select living animals at the facility with at least
one parent, the following can be used.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">focalAnimals</span> <span class="op">&lt;-</span> <span class="va">breederPed</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="op">!</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">breederPed</span><span class="op">$</span><span class="va">sire</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">breederPed</span><span class="op">$</span><span class="va">dam</span><span class="op">)</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">breederPed</span><span class="op">$</span><span class="va">exit</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/stringi/man/stri_join.html" class="external-link">stri_c</a></span><span class="op">(</span></span>
<span>  <span class="st">"There are "</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">focalAnimals</span><span class="op">)</span>,</span>
<span>  <span class="st">" animals in the vector _focalAnimals_."</span></span>
<span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>[1] “There are 327 animals in the vector <em>focalAnimals</em>.”</p>
<p>As can be seen, these animals have at least one parent and have not
left the facility.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">breederPed</span><span class="op">[</span><span class="va">breederPed</span><span class="op">$</span><span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span></span>
<span>  <span class="va">focalAnimals</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"id"</span>, <span class="st">"sire"</span>, <span class="st">"dam"</span>, <span class="st">"exit"</span><span class="op">)</span><span class="op">]</span><span class="op">[</span><span class="fl">1L</span><span class="op">:</span><span class="fl">10L</span>, <span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##          id   sire    dam exit</span></span>
<span><span class="co">## 1669 01QRQ4 VDBGDP TH7HTY &lt;NA&gt;</span></span>
<span><span class="co">## 1743 CLSVU6 ULV9M7 SUFWJI &lt;NA&gt;</span></span>
<span><span class="co">## 1887 1SPLS8 U9APLW 142GKP &lt;NA&gt;</span></span>
<span><span class="co">## 1934 5IAFMK U4YSS5 WVE6Y4 &lt;NA&gt;</span></span>
<span><span class="co">## 2072 HLQ9SY UI3RFL VEWC1E &lt;NA&gt;</span></span>
<span><span class="co">## 2234 XFWVVX U3MFSZ L4LM1F &lt;NA&gt;</span></span>
<span><span class="co">## 2337 6X6BG9 ENI6HX IUF0HC &lt;NA&gt;</span></span>
<span><span class="co">## 2377 B228Q6 UEUIRJ CBSIAA &lt;NA&gt;</span></span>
<span><span class="co">## 2378 B2CKHA ENI6HX WBFBR5 &lt;NA&gt;</span></span>
<span><span class="co">## 2383 BCJJKN UA379T JPVAT3 &lt;NA&gt;</span></span></code></pre>
<p>We indicate that these are the animals of interest by using the
<strong>setPopulation</strong> function. This function simply sets a
column named <em>population</em><a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;The &lt;em&gt;population&lt;/em&gt; column is created and added to
the pedigree object if it does not already exist.&lt;/p&gt;"><sup>2</sup></a> to the logical value of
<strong>TRUE</strong> if the row represents an animal in the list and
<strong>FALSE</strong> otherwise.</p>
<p>The first line of code below sets the <em>population</em> column and
the second counts the number of rows where the value was set to
<strong>TRUE</strong>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">breederPed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/setPopulation.html">setPopulation</a></span><span class="op">(</span>ped <span class="op">=</span> <span class="va">breederPed</span>, ids <span class="op">=</span> <span class="va">focalAnimals</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">breederPed</span><span class="op">[</span><span class="va">breederPed</span><span class="op">$</span><span class="va">population</span>, <span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 327</span></span></code></pre>
<p>The IDs used to populate the <em>population</em> flag can be used to
trim the pedigree so that it contains only those individuals who are in
the ID list or are ancestors of those individuals.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trimmedPed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/trimPedigree.html">trimPedigree</a></span><span class="op">(</span><span class="va">focalAnimals</span>, <span class="va">breederPed</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">breederPed</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 3694</span></span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">trimmedPed</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 704</span></span></code></pre>
<p>The <strong>trimPedigree</strong> function has the ability to remove
those ancestors that do not contribute genetic information.
Uninformative founders are those individuals who are parents of only one
individual and who have no parental information. (<em>Currently
genotypic information is ignored by
<strong>trimPedigree</strong></em>).</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trimmedPedInformative</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/trimPedigree.html">trimPedigree</a></span><span class="op">(</span><span class="va">focalAnimals</span>, <span class="va">breederPed</span>,</span>
<span>  removeUninformative <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">trimmedPedInformative</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 509</span></span></code></pre>
<p>We can find all of the animals that are in the trimmed pedigree but
are not focal animals.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nonfocalInTrimmedPed</span> <span class="op">&lt;-</span> <span class="va">trimmedPed</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="op">!</span><span class="va">trimmedPed</span><span class="op">$</span><span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">focalAnimals</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">nonfocalInTrimmedPed</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 377</span></span></code></pre>
<p>We can see which of these 377 are and are not parents. We will first
make sure we have all of the parents by getting our list of parents from
the entire pedigree. We then demonstrate that they are all in the
trimmed pedigree.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">allFocalParents</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="va">breederPed</span><span class="op">$</span><span class="va">sire</span><span class="op">[</span><span class="va">breederPed</span><span class="op">$</span><span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">focalAnimals</span><span class="op">]</span>,</span>
<span>  <span class="va">breederPed</span><span class="op">$</span><span class="va">dam</span><span class="op">[</span><span class="va">breederPed</span><span class="op">$</span><span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">focalAnimals</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="va">trimmedFocalParents</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="va">trimmedPed</span><span class="op">$</span><span class="va">sire</span><span class="op">[</span><span class="va">trimmedPed</span><span class="op">$</span><span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">focalAnimals</span><span class="op">]</span>,</span>
<span>  <span class="va">trimmedPed</span><span class="op">$</span><span class="va">dam</span><span class="op">[</span><span class="va">trimmedPed</span><span class="op">$</span><span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">focalAnimals</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.equal.html" class="external-link">all.equal</a></span><span class="op">(</span><span class="va">allFocalParents</span>, <span class="va">trimmedFocalParents</span><span class="op">)</span> <span class="co"># Are the IDs the same?</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<p>However, not all of the animals in the trimmed pedigree are either
the focal animals or their parents. They are more distant ancestors as
we will show.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">notFocalNotParent</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">trimmedPed</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="op">!</span><span class="va">trimmedPed</span><span class="op">$</span><span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">focalAnimals</span>, <span class="va">allFocalParents</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">notFocalNotParent</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 187</span></span></code></pre>
<p>Since the trimming process is supposed to retain the focal animals
and their ancestors, we will leave it as an exercise for you to
demonstrate that at least some of the remaining animals are grandparents
of the focal animals. <em>Hint: there are 490 grandparents in both the
trimmed and the complete pedigree</em>.</p>
<p>As you can see from the number of rows in the full pedigree (3694)
versus the trimmed pedigree (704), trimmed pedigrees can be much
smaller. Of the additional 377 animals, 182 provide genetic information
while the others (195) are genetically uninformative.</p>
<p>As is shown below only 4 (0ZX29Q, 1QBKW9, 5PWJ0G, and Y3CJ5A) living
animals are still in the colony but are not in the trimmed pedigree.<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;All animals within the colony have a known birth date.&lt;/p&gt;"><sup>3</sup></a></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">unknownBirth</span> <span class="op">&lt;-</span> <span class="va">breederPed</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">breederPed</span><span class="op">$</span><span class="va">birth</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">knownExit</span> <span class="op">&lt;-</span> <span class="va">breederPed</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">breederPed</span><span class="op">$</span><span class="va">exit</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">unknownBirthKnownExit</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">breederPed</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">breederPed</span><span class="op">$</span><span class="va">birth</span><span class="op">)</span> <span class="op">|</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">breederPed</span><span class="op">$</span><span class="va">exit</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">knownPed</span> <span class="op">&lt;-</span> <span class="va">breederPed</span><span class="op">[</span><span class="op">!</span><span class="va">breederPed</span><span class="op">$</span><span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">unknownBirthKnownExit</span>, <span class="op">]</span></span>
<span><span class="va">otherIds</span> <span class="op">&lt;-</span> <span class="va">knownPed</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="op">!</span><span class="va">knownPed</span><span class="op">$</span><span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">trimmedPed</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">trimmedPed</span><span class="op">$</span><span class="va">exit</span><span class="op">)</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/stringi/man/stri_join.html" class="external-link">stri_c</a></span><span class="op">(</span></span>
<span>  <span class="st">"The living animals in the pedigree that are not in the trimmed "</span>,</span>
<span>  <span class="st">"pedigree are "</span>, <span class="fu"><a href="../reference/get_and_or_list.html">get_and_or_list</a></span><span class="op">(</span><span class="va">otherIds</span><span class="op">)</span>, <span class="st">"."</span></span>
<span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>[1] “The living animals in the pedigree that are not in the trimmed
pedigree are 0ZX29Q, 1QBKW9, 5PWJ0G, and Y3CJ5A.”</p>
</div>
<div class="section level2">
<h2 id="age-sex-pyramid-plot">Age Sex Pyramid Plot<a class="anchor" aria-label="anchor" href="#age-sex-pyramid-plot"></a>
</h2>
<p>You can examine the population structure using an age-sex pyramid
plot with a single function. We will limit our view to just the focal
animals and their living relatives. This is appropriate for colony
management because in addition to the genetic diversity we seek, we have
to remain cognizant of the age and sex distributions within the colonies
we manage.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getPyramidPlot.html">getPyramidPlot</a></span><span class="op">(</span>ped <span class="op">=</span> <span class="va">trimmedPed</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">trimmedPed</span><span class="op">$</span><span class="va">exit</span><span class="op">)</span>, <span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="a2interactive_files/figure-html/plot-focal-age-sex-pyramid-1.png" alt="Age Sex Pyramid Plot." width="700"></p>
<pre><code><span><span class="co">## 45 45</span></span></code></pre>
<pre><code><span><span class="co">## [1] 5.1 4.1 4.1 2.1</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="genetic-value-analysis">Genetic Value Analysis<a class="anchor" aria-label="anchor" href="#genetic-value-analysis"></a>
</h2>
<p>Your genetic value analysis must be carefully performed. The next
three commands set up the entire pedigree for analysis. The first of
these three commands set all of the pedigree members to be part of the
population of interest by setting the <em>population</em> column to
<strong>TRUE</strong> for all individuals.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ped</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/setPopulation.html">setPopulation</a></span><span class="op">(</span><span class="va">breederPed</span>, <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<p>Note that a new pedigree object (<strong>ped</strong>) is being
created.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">probands</span> <span class="op">&lt;-</span> <span class="va">ped</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="va">ped</span><span class="op">$</span><span class="va">population</span><span class="op">]</span></span>
<span><span class="va">ped</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/trimPedigree.html">trimPedigree</a></span><span class="op">(</span><span class="va">probands</span>, <span class="va">ped</span>,</span>
<span>  removeUninformative <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  addBackParents <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The arguments to <strong>reportGV</strong> are all optional except
for <em>ped</em>, but you may often want to non-default values.</p>
<ul>
<li><p><strong>ped</strong> Pedigree information in data.frame
format</p></li>
<li><p><strong>guIter</strong> Integer indicating the number of
iterations for the gene-drop analysis. Default is 5000
iterations</p></li>
<li><p><strong>guThresh</strong> Integer indicating the threshold number
of animals for defining a unique allele. Default considers an allele
“unique” if it is found in only 1 animal.</p></li>
<li><p><strong>pop</strong> Character vector with animal IDs to consider
as the population of interest. The default is NULL.</p></li>
<li><p><strong>byID</strong> Logical variable of length 1 that is passed
through to eventually be used by alleleFreq(), which calculates the
count of each allele in the provided vector. If byID is TRUE and ids are
provided, the function will only count the unique alleles for an
individual (homozygous alleles will be counted as 1).</p></li>
</ul>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">geneticValue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/reportGV.html">reportGV</a></span><span class="op">(</span><span class="va">ped</span>,</span>
<span>  guIter <span class="op">=</span> <span class="fl">50L</span>,</span>
<span>  guThresh <span class="op">=</span> <span class="fl">3L</span>,</span>
<span>  byID <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  updateProgress <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">geneticValue</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## The genetic value report </span></span>
<span><span class="co">## Individuals in Pedigree: 3694 </span></span>
<span><span class="co">## Male Founders: 141</span></span>
<span><span class="co">## Female Founders: 122</span></span>
<span><span class="co">## Total Founders: 263 </span></span>
<span><span class="co">## Founder Equivalents: 241.84 </span></span>
<span><span class="co">## Founder Genome Equivalents: 164.01 </span></span>
<span><span class="co">## Live Offspring: 4052 </span></span>
<span><span class="co">## High Value Individuals: 2947 </span></span>
<span><span class="co">## Low Value Individuals: 747</span></span></code></pre>
<p>What happens if we limit our analysis to the trimmed pedigree?
Remember that the trimmed pedigree still contains all of the genetic
information that the full pedigree has for the focal animals.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trimmedGeneticValue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/reportGV.html">reportGV</a></span><span class="op">(</span><span class="va">trimmedPed</span>,</span>
<span>  guIter <span class="op">=</span> <span class="fl">50L</span>,</span>
<span>  guThresh <span class="op">=</span> <span class="fl">3L</span>,</span>
<span>  byID <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  updateProgress <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">trimmedGeneticValue</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## The genetic value report </span></span>
<span><span class="co">## Individuals in Pedigree: 327 </span></span>
<span><span class="co">## Male Founders: 3</span></span>
<span><span class="co">## Female Founders: 17</span></span>
<span><span class="co">## Total Founders: 20 </span></span>
<span><span class="co">## Founder Equivalents: 109.67 </span></span>
<span><span class="co">## Founder Genome Equivalents: 47.49 </span></span>
<span><span class="co">## Live Offspring: 321 </span></span>
<span><span class="co">## High Value Individuals: 226 </span></span>
<span><span class="co">## Low Value Individuals: 101</span></span></code></pre>
<p>It is clear that limiting your analysis to the animals of interest
reduces the effort required to examine the animals of interest.</p>
<div class="section level3">
<h3 id="detailed-look-at-the-genetic-value-report-object">Detailed look at the Genetic Value Report object<a class="anchor" aria-label="anchor" href="#detailed-look-at-the-genetic-value-report-object"></a>
</h3>
<p>The names of the object within the genetic value report object
(<em>trimmedGeneticValue</em>) can be listed as shown in the next line
of code.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">trimmedGeneticValue</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "report"          "kinship"         "gu"              "fe"             </span></span>
<span><span class="co">##  [5] "fg"              "maleFounders"    "femaleFounders"  "nMaleFounders"  </span></span>
<span><span class="co">##  [9] "nFemaleFounders" "total"</span></span></code></pre>
<p>The <em>report</em> object (an R dataframe) can in-turn be
examined.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">trimmedGeneticValue</span><span class="op">$</span><span class="va">report</span><span class="op">)</span> <span class="co">## column names</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "id"              "sex"             "age"             "birth"          </span></span>
<span><span class="co">##  [5] "exit"            "population"      "origin"          "indivMeanKin"   </span></span>
<span><span class="co">##  [9] "zScores"         "gu"              "totalOffspring"  "livingOffspring"</span></span>
<span><span class="co">## [13] "value"           "rank"</span></span></code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">trimmedGeneticValue</span><span class="op">$</span><span class="va">report</span><span class="op">)</span> <span class="co">## Number of rows</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 327</span></span></code></pre>
<p>The report is more conveniently used as a separate object. The next
section of code rounds some of the numerical values and converts all
columns to characters for display as a table where only the first 10
lines are included.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rpt</span> <span class="op">&lt;-</span> <span class="va">trimmedGeneticValue</span><span class="op">[[</span><span class="st">"report"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">rpt</span><span class="op">$</span><span class="va">indivMeanKin</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">rpt</span><span class="op">$</span><span class="va">indivMeanKin</span>, <span class="fl">5L</span><span class="op">)</span></span>
<span><span class="va">rpt</span><span class="op">$</span><span class="va">zScores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">rpt</span><span class="op">$</span><span class="va">zScores</span>, <span class="fl">2L</span><span class="op">)</span></span>
<span><span class="va">rpt</span><span class="op">$</span><span class="va">gu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">rpt</span><span class="op">$</span><span class="va">gu</span>, <span class="fl">5L</span><span class="op">)</span></span>
<span><span class="va">rpt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/toCharacter.html">toCharacter</a></span><span class="op">(</span><span class="va">rpt</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">rpt</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/headerDisplayNames.html">headerDisplayNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">rpt</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">rpt</span><span class="op">[</span><span class="fl">1L</span><span class="op">:</span><span class="fl">10L</span>, <span class="op">]</span><span class="op">)</span> <span class="co"># needs more work for display purposes.</span></span></code></pre></div>
<table style="width:100%;" class="table">
<colgroup>
<col width="3%">
<col width="1%">
<col width="7%">
<col width="5%">
<col width="4%">
<col width="11%">
<col width="3%">
<col width="11%">
<col width="11%">
<col width="10%">
<col width="7%">
<col width="8%">
<col width="8%">
<col width="2%">
</colgroup>
<thead><tr class="header">
<th align="left">Ego ID</th>
<th align="left">Sex</th>
<th align="right">Age (in years)</th>
<th align="left">Birth Date</th>
<th align="left">Exit Date</th>
<th align="left">Breeding Colony Member</th>
<th align="left">Origin</th>
<th align="right">Individual Mean Kinship</th>
<th align="right">Z-score (Mean Kinship)</th>
<th align="right">Genome Uniqueness (%)</th>
<th align="right">Total Offspring</th>
<th align="right">Living Offspring</th>
<th align="left">Value Designation</th>
<th align="right">Rank</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">KZM9RB</td>
<td align="left">M</td>
<td align="right">30.1</td>
<td align="left">1989-05-03</td>
<td align="left">NA</td>
<td align="left">TRUE</td>
<td align="left"></td>
<td align="right">0.00329</td>
<td align="right">-1.90</td>
<td align="right">92</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">High Value</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">CLSVU6</td>
<td align="left">F</td>
<td align="right">23.9</td>
<td align="left">1995-08-02</td>
<td align="left">NA</td>
<td align="left">TRUE</td>
<td align="left"></td>
<td align="right">0.00287</td>
<td align="right">-1.97</td>
<td align="right">90</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="left">High Value</td>
<td align="right">2</td>
</tr>
<tr class="odd">
<td align="left">1SPLS8</td>
<td align="left">F</td>
<td align="right">7.9</td>
<td align="left">2011-07-26</td>
<td align="left">NA</td>
<td align="left">TRUE</td>
<td align="left"></td>
<td align="right">0.00373</td>
<td align="right">-1.83</td>
<td align="right">84</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">High Value</td>
<td align="right">3</td>
</tr>
<tr class="even">
<td align="left">WK89I9</td>
<td align="left">F</td>
<td align="right">21.1</td>
<td align="left">1998-05-26</td>
<td align="left">NA</td>
<td align="left">TRUE</td>
<td align="left"></td>
<td align="right">0.00582</td>
<td align="right">-1.49</td>
<td align="right">80</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">High Value</td>
<td align="right">4</td>
</tr>
<tr class="odd">
<td align="left">8YP6PA</td>
<td align="left">M</td>
<td align="right">5.0</td>
<td align="left">2014-07-04</td>
<td align="left">NA</td>
<td align="left">TRUE</td>
<td align="left"></td>
<td align="right">0.00485</td>
<td align="right">-1.65</td>
<td align="right">77</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">High Value</td>
<td align="right">5</td>
</tr>
<tr class="even">
<td align="left">01QRQ4</td>
<td align="left">F</td>
<td align="right">18.2</td>
<td align="left">2001-04-04</td>
<td align="left">NA</td>
<td align="left">TRUE</td>
<td align="left"></td>
<td align="right">0.00373</td>
<td align="right">-1.83</td>
<td align="right">74</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">High Value</td>
<td align="right">6</td>
</tr>
<tr class="odd">
<td align="left">IZDV8K</td>
<td align="left">M</td>
<td align="right">7.7</td>
<td align="left">2011-09-29</td>
<td align="left">NA</td>
<td align="left">TRUE</td>
<td align="left"></td>
<td align="right">0.00480</td>
<td align="right">-1.66</td>
<td align="right">74</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">High Value</td>
<td align="right">7</td>
</tr>
<tr class="even">
<td align="left">R6HV9A</td>
<td align="left">M</td>
<td align="right">22.1</td>
<td align="left">1997-05-13</td>
<td align="left">NA</td>
<td align="left">TRUE</td>
<td align="left"></td>
<td align="right">0.00625</td>
<td align="right">-1.42</td>
<td align="right">73</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">High Value</td>
<td align="right">8</td>
</tr>
<tr class="odd">
<td align="left">CFD12A</td>
<td align="left">M</td>
<td align="right">20.8</td>
<td align="left">1998-08-25</td>
<td align="left">NA</td>
<td align="left">TRUE</td>
<td align="left"></td>
<td align="right">0.00509</td>
<td align="right">-1.61</td>
<td align="right">71</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">High Value</td>
<td align="right">9</td>
</tr>
<tr class="even">
<td align="left">3MMZD4</td>
<td align="left">M</td>
<td align="right">12.2</td>
<td align="left">2007-03-24</td>
<td align="left">NA</td>
<td align="left">TRUE</td>
<td align="left"></td>
<td align="right">0.00536</td>
<td align="right">-1.57</td>
<td align="right">70</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">High Value</td>
<td align="right">10</td>
</tr>
</tbody>
</table>
<p>We start the next lines of code by getting a fresh copy of the
genetic value report since we changed all of the numeric values to
characters in the last section to print the table. These lines
demonstrate one way of extracting the component objects from the
<em>genetic value</em> object.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rpt</span> <span class="op">&lt;-</span> <span class="va">trimmedGeneticValue</span><span class="op">[[</span><span class="st">"report"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">kmat</span> <span class="op">&lt;-</span> <span class="va">trimmedGeneticValue</span><span class="op">[[</span><span class="st">"kinship"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">f</span> <span class="op">&lt;-</span> <span class="va">trimmedGeneticValue</span><span class="op">[[</span><span class="st">"total"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">mf</span> <span class="op">&lt;-</span> <span class="va">trimmedGeneticValue</span><span class="op">[[</span><span class="st">"maleFounders"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">ff</span> <span class="op">&lt;-</span> <span class="va">trimmedGeneticValue</span><span class="op">[[</span><span class="st">"femaleFounders"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">nmf</span> <span class="op">&lt;-</span> <span class="va">trimmedGeneticValue</span><span class="op">[[</span><span class="st">"nMaleFounders"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">nff</span> <span class="op">&lt;-</span> <span class="va">trimmedGeneticValue</span><span class="op">[[</span><span class="st">"nFemaleFounders"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">fe</span> <span class="op">&lt;-</span> <span class="va">trimmedGeneticValue</span><span class="op">[[</span><span class="st">"fe"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">fg</span> <span class="op">&lt;-</span> <span class="va">trimmedGeneticValue</span><span class="op">[[</span><span class="st">"fg"</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p>It is informative to examine the distribution of <em>genetic
uniqueness</em>, <em>mean kinship</em>, and <em>z-scores</em>
(normalized <em>mean kinship</em> values).</p>
<p>Creation of the boxplot for the <em>genetic uniqueness</em> values is
shown below.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gu</span> <span class="op">&lt;-</span> <span class="va">rpt</span><span class="op">[</span>, <span class="st">"gu"</span><span class="op">]</span></span>
<span><span class="va">guBox</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>gu <span class="op">=</span> <span class="va">gu</span><span class="op">)</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="st">""</span>, y <span class="op">=</span> <span class="va">gu</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_boxplot</span><span class="op">(</span></span>
<span>    color <span class="op">=</span> <span class="st">"darkblue"</span>,</span>
<span>    fill <span class="op">=</span> <span class="st">"lightblue"</span>,</span>
<span>    notch <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co">#| fig.alt: &gt;</span></span>
<span>    <span class="co">#|   Histogram of time between eruptions for Old Faithful.</span></span>
<span>    <span class="co">#|   It is a bimodal distribution with peaks at 50-55 and</span></span>
<span>    <span class="co">#|   80-90 minutes.</span></span>
<span></span>
<span>    outlier.color <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>    outlier.shape <span class="op">=</span> <span class="fl">1L</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">coord_flip</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ylab</span><span class="op">(</span><span class="st">"Score"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Genetic Uniqueness"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">guBox</span><span class="op">)</span></span></code></pre></div>
<p><img src="a2interactive_files/figure-html/genetic-uniqueness-boxplot-1.png" alt="Genetic Uniqueness Box Plot." width="700"></p>
<p>Extraction of the individual <em>mean kinship</em> values and their
corresponding z-scores is shown in the next code chunk.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mk</span> <span class="op">&lt;-</span> <span class="va">rpt</span><span class="op">[</span>, <span class="st">"indivMeanKin"</span><span class="op">]</span></span>
<span><span class="va">zs</span> <span class="op">&lt;-</span> <span class="va">rpt</span><span class="op">[</span>, <span class="st">"zScores"</span><span class="op">]</span></span></code></pre></div>
<p>Creation of boxplots for the <em>mean kinship</em> and
<em>z-scores</em> is left as an exercise.</p>
</div>
</div>
<div class="section level2">
<h2 id="breeding-group-formation">Breeding Group Formation<a class="anchor" aria-label="anchor" href="#breeding-group-formation"></a>
</h2>
<p>The primary purpose of <strong>nprcgenekeepr</strong> is to form
breeding groups according to our best information regarding maintaining
the genetic characteristics we desire and the realities associated with
other animal husbandry needs.</p>
<p>There are several options you must consider when forming groups using
<strong>nprcgenekeepr</strong>, which we will examine using code
below.</p>
<ul>
<li>Animals used to form groups
<ul>
<li>
<em>high-value</em>: Randomly select from only high-value animals in
genetic value analysis</li>
<li>
<em>all</em>: Randomly select from all animals in genetic value
analysis</li>
<li>
<em>candidates</em>: Use candidate animals entered below to form
groups</li>
</ul>
</li>
<li>Sex ratio
<ul>
<li>Randomly assign animals without regard to sex</li>
<li>Use a harem structure with one breeding male per group</li>
<li>Specify the sex ratio between 0.5 and 10 (F/M)</li>
</ul>
</li>
<li>Whether or not to pre-populate (seed) groups with animals of your
choice</li>
<li>Number of groups to be formed</li>
<li>Whether or not to ignore females at or above the minimum parent
age</li>
<li>Number of simulations used to search for the optimal group
makeup</li>
<li>Whether or not kinship coefficients are to be included in
results</li>
</ul>
<p>You decisions regarding each of the above options are expressed in a
call to the function <strong>groupAddAssign</strong>. A complete
description of the function and its arguments is available using the
code shown below.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="va">groupAddAssign</span></span></code></pre></div>
<p>Below is the descriptions of the function parameters extracted from
the documentation near the time this tutorial was prepared.</p>
<ul>
<li><p><strong>candidates</strong> Character vector of IDs of the
animals available for use in forming the groups. The animals that may be
present in currentGroups are not included within candidates.</p></li>
<li><p><strong>currentGroups</strong> List of character vectors of IDs
of animals currently assigned to groups. Defaults to a list with
character(0) in each sub-list element (one for each group being formed)
assuming no groups are pre-populated.</p></li>
<li><p><strong>kmat</strong> Numeric matrix of pairwise kinship values.
Rows and columns are named with animal IDs.</p></li>
<li><p><strong>ped</strong> Dataframe that is the ‘Pedigree’. It
contains pedigree information including the IDs listed in
candidates.</p></li>
<li><p><strong>threshold</strong> Numeric value indicating the minimum
kinship level to be considered in group formation. Pairwise kinship
below this level will be ignored. The default values is
0.015625.</p></li>
<li><p><strong>ignore</strong> List of character vectors representing
the sex combinations to be ignored. If provided, the vectors in the list
specify if pairwise kinship should be ignored between certain sexes.
Default is to ignore all pairwise kinship between females.</p></li>
<li><p><strong>minAge</strong> Integer value indicating the minimum age
to consider in group formation. Pairwise kinships involving an animal of
this age or younger will be ignored. Default is 1 year.</p></li>
<li><p><strong>iter</strong> Integer indicating the number of times to
perform the random group formation process. Default value is 1000
iterations.</p></li>
<li><p><strong>numGp</strong> Integer value indicating the number of
groups that should be formed from the list of IDs. Default is
1.</p></li>
<li><p><strong>updateProgress</strong> Function or NULL. If this
function is defined, it will be called during each iteration to update a
shiny::Progress object.</p></li>
<li><p><strong>harem</strong> Logical variable when set to TRUE, the
formed groups have a single male at least minAge old.</p></li>
<li><p><strong>sexRatio</strong> Numeric value indicating the ratio of
females to males x from 0.5 to 20 by increments of 0.5.</p></li>
<li><p><strong>withKin</strong> Logical variable when set to TRUE, the
kinship matrix for the group is returned along with the group and score.
Defaults to not return the kinship matrix. This maintains compatibility
with earlier versions.</p></li>
</ul>
<p>We will use the <em>trimmedPed</em> pedigree in our code.</p>
<p>For illustration purposes we are going to keep the numbers of
candidates, groups, and iterations fairly small.</p>
<p>We will get first some animal IDs to use for our candidates by
selecting animals at least 2 years old at the time this pedigree was
sampled (01-01-2015).</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">candidates</span> <span class="op">&lt;-</span> <span class="va">trimmedPed</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="va">trimmedPed</span><span class="op">$</span><span class="va">birth</span> <span class="op">&lt;</span> <span class="fu"><a href="https://rdrr.io/r/base/as.Date.html" class="external-link">as.Date</a></span><span class="op">(</span><span class="st">"2013-01-01"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">trimmedPed</span><span class="op">$</span><span class="va">birth</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">trimmedPed</span><span class="op">$</span><span class="va">exit</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">trimmedPed</span><span class="op">$</span><span class="va">sex</span><span class="op">[</span><span class="va">trimmedPed</span><span class="op">$</span><span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">candidates</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##   F   M   H   U </span></span>
<span><span class="co">## 184  96   0   0</span></span></code></pre>
<p>Our candidates are made up of 184 females and 96 males. The
parameters <strong>currentGroups</strong>, <strong>threshold</strong>,
<strong>ignore</strong>, <strong>minAge</strong>,
<strong>sexRatio</strong>, <strong>withKin</strong>, and
<strong>updateProgress</strong> are allowed to take their default
values. The setting of the <strong>sexRatio</strong> parameter to 0 is
ignored in the following call of the <strong>groupAddAssign</strong>
function. This is consistent with the a value of 0 making little since
in a breeding colony.</p>
<p>The empty seventh group at the bottom is evidence that all of the
candidate animals could be placed in a group without exceeding the
default value of 0.015625.</p>
<div class="section level3">
<h3 id="harems">Harems<a class="anchor" aria-label="anchor" href="#harems"></a>
</h3>
<p>The following group assignments will be forming harem groups. This is
done by setting <strong>harem</strong> to . Setting <code>iter</code> to
100 or more will increase optimal composition of breeding groups</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">haremGrp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/groupAddAssign.html">groupAddAssign</a></span><span class="op">(</span></span>
<span>  candidates <span class="op">=</span> <span class="va">candidates</span>,</span>
<span>  kmat <span class="op">=</span> <span class="va">trimmedGeneticValue</span><span class="op">[[</span><span class="st">"kinship"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  ped <span class="op">=</span> <span class="va">trimmedPed</span>,</span>
<span>  iter <span class="op">=</span> <span class="fl">10L</span>,</span>
<span>  numGp <span class="op">=</span> <span class="fl">6L</span>,</span>
<span>  harem <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">haremGrp</span><span class="op">$</span><span class="va">group</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span>
<span><span class="co">##  [1] "S63QDN" "TQEMY6" "XYRDKV" "ZATMEE" "AW400C" "AR5U44" "CHK1ZX" "414N7M"</span></span>
<span><span class="co">##  [9] "EZ2F8A" "WLMGS1" "MYUMMX" "DHNQ1W" "1KJ2MG" "KZY6PD" "0SGJ12" "CS23RV"</span></span>
<span><span class="co">## [17] "AP1YLW" "38K2SR" "VWC5ZH" "GTLA8R" "1CIRC9" "YLRNIK" "5KWNMZ" "83HQBN"</span></span>
<span><span class="co">## [25] "465ERA" "Y6DB6L" "7NE2UT" "EX5K0S" "MTCAIG"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span>
<span><span class="co">##  [1] "CHJ9D2" "T3QPW5" "XEC0M5" "YFCIHJ" "S222R3" "BKWE4D" "S3EBGZ" "8JUUJ9"</span></span>
<span><span class="co">##  [9] "72LYDE" "6X6BG9" "CRPXY7" "FB5L3N" "87AQLF" "QCENKM" "MX4J7G" "BCJJKN"</span></span>
<span><span class="co">## [17] "WKY2SZ" "QRZK48" "01QRQ4" "SCFSBF" "AFZKBS" "WI38KZ" "DH9WJQ" "BS3RLE"</span></span>
<span><span class="co">## [25] "Y0TCYX" "30J3CQ" "MFKT9C"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[3]]</span></span>
<span><span class="co">##  [1] "FX9E4X" "B228Q6" "CMMUKU" "7B9CA6" "N4NV8B" "MH88T6" "I5CI33" "M9PVG5"</span></span>
<span><span class="co">##  [9] "1SSCJC" "WNEAS6" "46ZHKN" "967Y3D" "2F6J3U" "E5Q33K" "C18V6I" "3YJIMV"</span></span>
<span><span class="co">## [17] "N5QBWD" "AIHJ8Z" "1CZM30" "DI4AHD" "YTJ2UL" "SHG3RB" "AZ3L0D" "X694YR"</span></span>
<span><span class="co">## [25] "1FAZ0K" "DRXMW4" "D4B0RM" "Z904TJ" "0XTZQ1" "GCBYDW" "FL170P" "50D77I"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[4]]</span></span>
<span><span class="co">##  [1] "Z25D52" "NK802Y" "IH1KPA" "PJ72W1" "Q8U9LB" "B134XZ" "RJ4JPC" "MKY9TK"</span></span>
<span><span class="co">##  [9] "S7IWWA" "QW2Z3R" "D9P18Y" "5W621W" "3GECJJ" "S056D5" "SH3FB7" "HE0SCR"</span></span>
<span><span class="co">## [17] "H2J6UA" "TYEWF1" "WTE53B" "GIIEUD" "D33J06" "PU7RSG" "3SKITJ" "6F9FB8"</span></span>
<span><span class="co">## [25] "W6MDVK" "J3F6PD" "R5AYJK" "Q7U139" "7RA57Q" "5EDLL7" "RVHVTZ" "WJXIH9"</span></span>
<span><span class="co">## [33] "PVY432"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[5]]</span></span>
<span><span class="co">##  [1] "YDRD81" "G58RGY" "5ERY5Z" "1SPLS8" "6KWVRI" "321LLB" "K3TNHP" "TEACA3"</span></span>
<span><span class="co">##  [9] "2Z4YLY" "FJS7RQ" "9MG040" "JLFKV8" "N79QXB" "5EDIEE" "13B1QL" "QWKFBH"</span></span>
<span><span class="co">## [17] "QCA36T" "WK89I9" "I8ABC7" "LYSLPP" "MB6NYQ" "1GF3GM" "DKIM6U" "99BMJW"</span></span>
<span><span class="co">## [25] "KX0RJ3" "0HYZ23" "QQMBT1" "92UG4N" "GAS52W" "DPXEQE" "G25E3F" "PYPM1W"</span></span>
<span><span class="co">## [33] "EMV4P6" "TXZUKC" "S5H1GC" "1QVS67" "CLSVU6" "LVYYNY" "5BPBUI"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[6]]</span></span>
<span><span class="co">##  [1] "XY2CK7" "F45799" "NN3GDQ" "KEA4QG" "LS184H" "9P0DES" "MPIQ4N" "B1WVCN"</span></span>
<span><span class="co">##  [9] "W5WIRP" "3DTD2N" "5IAFMK" "DCJJYS" "PI4VHT" "0X4W26" "FG0SFA" "7ZNY75"</span></span>
<span><span class="co">## [17] "G8MCV7" "ZQXZYB" "PBAFJF" "ZPS15A" "ILVQVB" "ZH3YG1" "Q17CG3" "W0GUKI"</span></span>
<span><span class="co">## [25] "F7I2ED" "AR17R5" "XFWVVX" "1VP3UC" "0IIAEN" "ESUIAF"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[7]]</span></span>
<span><span class="co">## [1] NA</span></span></code></pre>
<p>We can identify and list the males in each group with the following
code.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">haremGrp</span><span class="op">$</span><span class="va">group</span>, <span class="kw">function</span><span class="op">(</span><span class="va">ids</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">ids</span><span class="op">[</span><span class="va">ids</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">trimmedPed</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="va">trimmedPed</span><span class="op">$</span><span class="va">sex</span> <span class="op">==</span> <span class="st">"M"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span>
<span><span class="co">## [1] "S63QDN"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span>
<span><span class="co">## [1] "CHJ9D2"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[3]]</span></span>
<span><span class="co">## [1] "FX9E4X"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[4]]</span></span>
<span><span class="co">## [1] "Z25D52"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[5]]</span></span>
<span><span class="co">## [1] "YDRD81"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[6]]</span></span>
<span><span class="co">## [1] "XY2CK7"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[7]]</span></span>
<span><span class="co">## logical(0)</span></span></code></pre>
<p>It is easy to notice that the male is listed first within each
breeding group.</p>
<p>We can also see the number of animals and the sex ratios created in
each group. Since these are harem groups the sex ratios are determined
by the number of animals in the individual groups.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lines</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">haremGrp</span><span class="op">$</span><span class="va">group</span>, <span class="kw">function</span><span class="op">(</span><span class="va">ids</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span></span>
<span>    <span class="st">"Count: "</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ids</span><span class="op">)</span>, <span class="st">" Sex Ratio: "</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="../reference/calculateSexRatio.html">calculateSexRatio</a></span><span class="op">(</span><span class="va">ids</span>, <span class="va">trimmedPed</span><span class="op">)</span>, <span class="fl">2</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">line</span> <span class="kw">in</span> <span class="va">lines</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">line</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Count: 29 Sex Ratio: 28"</span></span>
<span><span class="co">## [1] "Count: 27 Sex Ratio: 26"</span></span>
<span><span class="co">## [1] "Count: 32 Sex Ratio: 31"</span></span>
<span><span class="co">## [1] "Count: 33 Sex Ratio: 32"</span></span>
<span><span class="co">## [1] "Count: 39 Sex Ratio: 38"</span></span>
<span><span class="co">## [1] "Count: 30 Sex Ratio: 29"</span></span>
<span><span class="co">## [1] "Count: 1 Sex Ratio: Inf"</span></span></code></pre>
<p>Examination of this table shows that of the 184 females 156 are
included.</p>
</div>
<div class="section level3">
<h3 id="controlling-sex-ratios">Controlling Sex Ratios<a class="anchor" aria-label="anchor" href="#controlling-sex-ratios"></a>
</h3>
<p>The following group assignments will be forming harem groups. This is
done by setting <strong>harem</strong> to .</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sexRatioGrp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/groupAddAssign.html">groupAddAssign</a></span><span class="op">(</span></span>
<span>  candidates <span class="op">=</span> <span class="va">candidates</span>,</span>
<span>  kmat <span class="op">=</span> <span class="va">trimmedGeneticValue</span><span class="op">[[</span><span class="st">"kinship"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  ped <span class="op">=</span> <span class="va">trimmedPed</span>,</span>
<span>  iter <span class="op">=</span> <span class="fl">10L</span>,</span>
<span>  numGp <span class="op">=</span> <span class="fl">6L</span>,</span>
<span>  sexRatio <span class="op">=</span> <span class="fl">9.0</span></span>
<span><span class="op">)</span></span>
<span><span class="va">sexRatioGrp</span><span class="op">$</span><span class="va">group</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span>
<span><span class="co">##  [1] "SHG3RB" "8TV4MT" "QRZK48" "D9P18Y" "414N7M" "SH3FB7" "0XTZQ1" "3YJIMV"</span></span>
<span><span class="co">##  [9] "BKWE4D" "83HQBN" "Z904TJ" "6KWVRI" "NN3GDQ" "KZM9RB" "BS3RLE" "5EDIEE"</span></span>
<span><span class="co">## [17] "KEA4QG" "DRXMW4" "87AQLF" "2Z4YLY" "0X4W26" "ESUIAF" "1CIRC9" "WNEAS6"</span></span>
<span><span class="co">## [25] "HLQ9SY" "1SPLS8" "PU7RSG" "S7IWWA" "S222R3" "5BPBUI" "WLMGS1" "J3F6PD"</span></span>
<span><span class="co">## [33] "MPIQ4N" "WJXIH9" "G2GYST"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span>
<span><span class="co">##  [1] "Y6DB6L" "XY2CK7" "1KJ2MG" "XEC0M5" "DHNQ1W" "KZY6PD" "5KWNMZ" "B228Q6"</span></span>
<span><span class="co">##  [9] "5EDLL7" "Y0TCYX" "CRPXY7" "G8MCV7" "GIIEUD" "R6HV9A" "W6MDVK" "CS23RV"</span></span>
<span><span class="co">## [17] "1SSCJC" "01QRQ4" "MB6NYQ" "1CZM30" "WKY2SZ" "VWC5ZH" "PBAFJF" "Q7U139"</span></span>
<span><span class="co">## [25] "CFD12A" "MYUMMX"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[3]]</span></span>
<span><span class="co">##  [1] "XYRDKV" "5XVTVH" "QCA36T" "RVHVTZ" "3SKITJ" "WK89I9" "FG0SFA" "DI4AHD"</span></span>
<span><span class="co">##  [9] "WTE53B" "XFWVVX" "C18V6I" "MH88T6" "321LLB" "YHHVC7" "TQEMY6" "Q17CG3"</span></span>
<span><span class="co">## [17] "50D77I" "AFZKBS" "3DTD2N" "ZATMEE" "30J3CQ" "EMV4P6" "DPXEQE" "8JUUJ9"</span></span>
<span><span class="co">## [25] "K7900I"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[4]]</span></span>
<span><span class="co">##  [1] "MKY9TK" "2F1IV1" "1QVS67" "MTCAIG" "1GF3GM" "YFCIHJ" "W0GUKI" "465ERA"</span></span>
<span><span class="co">##  [9] "EX5K0S" "6F9FB8" "YTJ2UL" "PVY432" "D33J06" "IZDV8K" "99BMJW" "AZ3L0D"</span></span>
<span><span class="co">## [17] "R5AYJK" "N5QBWD" "NK802Y" "N79QXB" "46ZHKN" "WI38KZ" "72LYDE" "0HYZ23"</span></span>
<span><span class="co">## [25] "3MMZD4"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[5]]</span></span>
<span><span class="co">##  [1] "DH9WJQ" "XL658N" "QQMBT1" "X694YR" "AIHJ8Z" "7RA57Q" "LVYYNY" "0IIAEN"</span></span>
<span><span class="co">##  [9] "1VP3UC" "Q8U9LB" "GCBYDW" "MX4J7G" "FL170P" "T38W6H" "PI4VHT" "S5H1GC"</span></span>
<span><span class="co">## [17] "5ERY5Z" "JLFKV8" "ZQXZYB" "13B1QL" "7NE2UT" "GTLA8R" "PJ72W1" "DKIM6U"</span></span>
<span><span class="co">## [25] "LN1DLY" "EZ2F8A" "G25E3F" "KX0RJ3" "7ZNY75" "LS184H" "9MG040" "ZH3YG1"</span></span>
<span><span class="co">## [33] "N4NV8B" "B134XZ" "Z7NBA2"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[6]]</span></span>
<span><span class="co">##  [1] "38K2SR" "B2CKHA" "QCENKM" "AP1YLW" "CHK1ZX" "YLRNIK" "QWKFBH" "D4B0RM"</span></span>
<span><span class="co">##  [9] "GAS52W" "TXZUKC" "G58RGY" "967Y3D" "CLSVU6" "7D09WH" "1FAZ0K" "ILVQVB"</span></span>
<span><span class="co">## [17] "B1WVCN" "RJ4JPC" "T3QPW5" "S3EBGZ" "AR17R5" "LYSLPP" "F7I2ED" "FJS7RQ"</span></span>
<span><span class="co">## [25] "5PW7WT" "FB5L3N" "HE0SCR" "H2J6UA" "BCJJKN" "K3TNHP" "S056D5"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[7]]</span></span>
<span><span class="co">##   [1] "5IAFMK" "6X6BG9" "DCJJYS" "GDXWJ1" "JSAP3H" "TR5L57" "XC304E" "1E8KD1"</span></span>
<span><span class="co">##   [9] "5KFB90" "A6A1M1" "AEP5EG" "AW400C" "BW10CL" "CHJ9D2" "FTVE03" "IH1KPA"</span></span>
<span><span class="co">##  [17] "IRFJ09" "KXHGRH" "LMJWTN" "M9PVG5" "Q9LWGX" "RNQU14" "SCFSBF" "TEACA3"</span></span>
<span><span class="co">##  [25] "TYEWF1" "W5WIRP" "ZPS15A" "09LFE4" "2F6J3U" "3GECJJ" "3QHAFI" "55BPSE"</span></span>
<span><span class="co">##  [33] "7B9CA6" "8IG767" "9FRCIE" "ER464J" "FFGPS4" "FG6L7S" "NHWTJ9" "P7RBPI"</span></span>
<span><span class="co">##  [41] "TBCE78" "YI16QD" "4LHK19" "59NYZE" "5IYDXN" "6KLWVC" "80F2MI" "A98D7P"</span></span>
<span><span class="co">##  [49] "AZ4D19" "BTTHAJ" "CHSCFG" "EEGLWY" "FX9E4X" "G91ZM6" "I5CI33" "I8ABC7"</span></span>
<span><span class="co">##  [57] "J1R2EW" "LDND6J" "MFKT9C" "MQT080" "NSIC4I" "PHB6TE" "PYPM1W" "QRWYQZ"</span></span>
<span><span class="co">##  [65] "QW2Z3R" "RY1AZM" "WHQLH5" "WQUN84" "XX0GYV" "YP910X" "0SGJ12" "0V4SAC"</span></span>
<span><span class="co">##  [73] "0X1RZ9" "3YHBC1" "55VDSQ" "5W621W" "653J82" "6MEP2C" "76DIT4" "80KACX"</span></span>
<span><span class="co">##  [81] "92UG4N" "9P0DES" "B2YJJP" "CMMUKU" "E3JP0C" "E5Q33K" "F45799" "FLIZQI"</span></span>
<span><span class="co">##  [89] "GM371F" "MEUZ85" "PA9F3J" "SXSVEH" "TJN1AD" "WNKKW3" "XZH41H" "YDRD81"</span></span>
<span><span class="co">##  [97] "Z25D52" "ZDRSG0" "3P9BX6" "AR5U44" "DGZLV3" "S63QDN" "ZW2X4N"</span></span></code></pre>
<p>Again we can identify and list the males in each group with the
following code.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">sexRatioGrp</span><span class="op">$</span><span class="va">group</span>, <span class="kw">function</span><span class="op">(</span><span class="va">ids</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">ids</span><span class="op">[</span><span class="va">ids</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">trimmedPed</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="va">trimmedPed</span><span class="op">$</span><span class="va">sex</span> <span class="op">==</span> <span class="st">"M"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span>
<span><span class="co">## [1] "8TV4MT" "KZM9RB" "HLQ9SY" "G2GYST"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span>
<span><span class="co">## [1] "XY2CK7" "R6HV9A" "CFD12A"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[3]]</span></span>
<span><span class="co">## [1] "5XVTVH" "YHHVC7" "K7900I"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[4]]</span></span>
<span><span class="co">## [1] "2F1IV1" "IZDV8K" "3MMZD4"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[5]]</span></span>
<span><span class="co">## [1] "XL658N" "T38W6H" "LN1DLY" "Z7NBA2"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[6]]</span></span>
<span><span class="co">## [1] "B2CKHA" "7D09WH" "5PW7WT"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[7]]</span></span>
<span><span class="co">##  [1] "GDXWJ1" "JSAP3H" "TR5L57" "XC304E" "1E8KD1" "5KFB90" "A6A1M1" "AEP5EG"</span></span>
<span><span class="co">##  [9] "BW10CL" "CHJ9D2" "FTVE03" "IRFJ09" "KXHGRH" "LMJWTN" "Q9LWGX" "RNQU14"</span></span>
<span><span class="co">## [17] "09LFE4" "3QHAFI" "55BPSE" "8IG767" "9FRCIE" "ER464J" "FFGPS4" "FG6L7S"</span></span>
<span><span class="co">## [25] "NHWTJ9" "P7RBPI" "TBCE78" "YI16QD" "4LHK19" "59NYZE" "5IYDXN" "6KLWVC"</span></span>
<span><span class="co">## [33] "80F2MI" "A98D7P" "AZ4D19" "BTTHAJ" "CHSCFG" "EEGLWY" "FX9E4X" "G91ZM6"</span></span>
<span><span class="co">## [41] "J1R2EW" "LDND6J" "MQT080" "NSIC4I" "PHB6TE" "QRWYQZ" "RY1AZM" "WHQLH5"</span></span>
<span><span class="co">## [49] "WQUN84" "XX0GYV" "YP910X" "0V4SAC" "0X1RZ9" "3YHBC1" "55VDSQ" "653J82"</span></span>
<span><span class="co">## [57] "6MEP2C" "76DIT4" "80KACX" "B2YJJP" "E3JP0C" "FLIZQI" "GM371F" "MEUZ85"</span></span>
<span><span class="co">## [65] "PA9F3J" "SXSVEH" "TJN1AD" "WNKKW3" "XZH41H" "YDRD81" "Z25D52" "ZDRSG0"</span></span>
<span><span class="co">## [73] "3P9BX6" "DGZLV3" "S63QDN" "ZW2X4N"</span></span></code></pre>
<p>We can also see the number of animals and the sex ratios created in
each group.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lines</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">sexRatioGrp</span><span class="op">$</span><span class="va">group</span>, <span class="kw">function</span><span class="op">(</span><span class="va">ids</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span></span>
<span>    <span class="st">"Count: "</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ids</span><span class="op">)</span>, <span class="st">" Sex Ratio: "</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="../reference/calculateSexRatio.html">calculateSexRatio</a></span><span class="op">(</span><span class="va">ids</span>, <span class="va">trimmedPed</span><span class="op">)</span>, <span class="fl">2L</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">line</span> <span class="kw">in</span> <span class="va">lines</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">line</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Count: 35 Sex Ratio: 7.75"</span></span>
<span><span class="co">## [1] "Count: 26 Sex Ratio: 7.67"</span></span>
<span><span class="co">## [1] "Count: 25 Sex Ratio: 7.33"</span></span>
<span><span class="co">## [1] "Count: 25 Sex Ratio: 7.33"</span></span>
<span><span class="co">## [1] "Count: 35 Sex Ratio: 7.75"</span></span>
<span><span class="co">## [1] "Count: 31 Sex Ratio: 9.33"</span></span>
<span><span class="co">## [1] "Count: 103 Sex Ratio: 0.36"</span></span></code></pre>
<p>Examination of this table shows that of the 184 females 239 are
included.</p>
</div>
</div>
<div class="section level2">
<h2 id="pedigree-errors">Pedigree Errors<a class="anchor" aria-label="anchor" href="#pedigree-errors"></a>
</h2>
<p>As stated earlier you can see which types of errors are detected by
<em>qcStudbook</em> by looking at names returned by
<em>getEmptyErrorLst()</em> as shown below.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="../reference/getEmptyErrorLst.html">getEmptyErrorLst</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "failedDatabaseConnection" "missingColumns"          </span></span>
<span><span class="co">## [3] "invalidDateRows"          "suspiciousParents"       </span></span>
<span><span class="co">## [5] "femaleSires"              "maleDams"                </span></span>
<span><span class="co">## [7] "sireAndDam"               "duplicateIds"            </span></span>
<span><span class="co">## [9] "changedCols"</span></span></code></pre>
<p>Each is defined below.</p>
<table class="table">
<colgroup>
<col width="19%">
<col width="80%">
</colgroup>
<thead><tr class="header">
<th align="left">Error</th>
<th align="left">Definition</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">failedDatabaseConnection</td>
<td align="left">Database connection failed: configuration or
permissions are invalid</td>
</tr>
<tr class="even">
<td align="left">missingColumns</td>
<td align="left">Columns that must be within the pedigree file are
missing.</td>
</tr>
<tr class="odd">
<td align="left">invalidDateRows</td>
<td align="left">Values, which are supposed to be dates, cannot be
interpreted as a date.</td>
</tr>
<tr class="even">
<td align="left">suspiciousParents</td>
<td align="left">Parents were too young on the date of birth of to have
been the parent.</td>
</tr>
<tr class="odd">
<td align="left">femaleSires</td>
<td align="left">Individuals listed as female or hermaphroditic and as a
sire.</td>
</tr>
<tr class="even">
<td align="left">maleDams</td>
<td align="left">Individuals are listed as male and as a dam.</td>
</tr>
<tr class="odd">
<td align="left">sireAndDam</td>
<td align="left">Individuals who are listed as both a sire and a
dam.</td>
</tr>
<tr class="even">
<td align="left">duplicateIds</td>
<td align="left">IDs listed more than once.</td>
</tr>
<tr class="odd">
<td align="left">changedCols</td>
<td align="left">Columns that have been changed to conform to internal
naming conventions and what they were changed to.</td>
</tr>
</tbody>
</table>
<p>We are going to use the small imaginary pedigree listed below that
has multiple errors to discuss pedigree error detection and reporting.
First note the birth dates of ego_id <em>o4</em> (2006-04-13) and the
purported sire <em>s2</em> (2006-06-19). Note also the purported birth
date of the <em>d2</em> and the birth dates of her offspring. Obviously
dates or IDs may be incorrect.</p>
<p>This is the pedigree. <em>(We will discuss the column names
shortly.)</em></p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu">nprcgenekeepr</span><span class="fu">::</span><span class="va"><a href="../reference/pedOne.html">pedOne</a></span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">ego_id</th>
<th align="left">si re</th>
<th align="left">dam_id</th>
<th align="left">sex</th>
<th align="left">birth_date</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">s1</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">F</td>
<td align="left">2000-07-18</td>
</tr>
<tr class="even">
<td align="left">d1</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">M</td>
<td align="left">2003-04-13</td>
</tr>
<tr class="odd">
<td align="left">s2</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">M</td>
<td align="left">2006-06-19</td>
</tr>
<tr class="even">
<td align="left">d2</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">F</td>
<td align="left">2015-09-16</td>
</tr>
<tr class="odd">
<td align="left">o1</td>
<td align="left">s1</td>
<td align="left">d1</td>
<td align="left">F</td>
<td align="left">2015-02-04</td>
</tr>
<tr class="even">
<td align="left">o2</td>
<td align="left">s1</td>
<td align="left">d2</td>
<td align="left">F</td>
<td align="left">2009-03-17</td>
</tr>
<tr class="odd">
<td align="left">o3</td>
<td align="left">s2</td>
<td align="left">d2</td>
<td align="left">F</td>
<td align="left">2012-04-11</td>
</tr>
<tr class="even">
<td align="left">o4</td>
<td align="left">s2</td>
<td align="left">d2</td>
<td align="left">M</td>
<td align="left">2006-04-13</td>
</tr>
</tbody>
</table>
<p>If we try to convert this pedigree file into the standardized
studbook format, we are going to get an error message and the creation
of a file in the R sessions temporary directory that lists the records
that have generated the errors.</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pedOne</span> <span class="op">&lt;-</span> <span class="fu">nprcgenekeepr</span><span class="fu">::</span><span class="va"><a href="../reference/pedOne.html">pedOne</a></span> <span class="co"># put it in the local environment</span></span>
<span><span class="va">ped</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qcStudbook.html">qcStudbook</a></span><span class="op">(</span><span class="va">pedOne</span>, minParentAge <span class="op">=</span> <span class="fl">0.0</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Error in qcStudbook(pedOne, minParentAge = 0): Parents with low age at birth of offspring are listed in /tmp/RtmplYP5PH/lowParentAge.csv.</span></span></code></pre>
<p>The contents of <em>lowParentAge.csv</em> is shown below.</p>
<table class="table">
<colgroup>
<col width="4%">
<col width="6%">
<col width="3%">
<col width="4%">
<col width="13%">
<col width="15%">
<col width="6%">
<col width="13%">
<col width="13%">
<col width="9%">
<col width="8%">
</colgroup>
<thead><tr class="header">
<th align="left">dam</th>
<th align="left">sire</th>
<th align="left">id</th>
<th align="left">sex</th>
<th align="left">birth</th>
<th align="left">recordStatus</th>
<th align="left">exit</th>
<th align="left">sireBirth</th>
<th align="left">damBirth</th>
<th align="right">sireAge</th>
<th align="right">damAge</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">d2</td>
<td align="left">s1</td>
<td align="left">o2</td>
<td align="left">F</td>
<td align="left">2009-03-17</td>
<td align="left">original</td>
<td align="left">NA</td>
<td align="left">2000-07-18</td>
<td align="left">2015-09-16</td>
<td align="right">8.66</td>
<td align="right">-6.50</td>
</tr>
<tr class="even">
<td align="left">d2</td>
<td align="left">s2</td>
<td align="left">o3</td>
<td align="left">F</td>
<td align="left">2012-04-11</td>
<td align="left">original</td>
<td align="left">NA</td>
<td align="left">2006-06-19</td>
<td align="left">2015-09-16</td>
<td align="right">5.81</td>
<td align="right">-3.43</td>
</tr>
<tr class="odd">
<td align="left">d2</td>
<td align="left">s2</td>
<td align="left">o4</td>
<td align="left">M</td>
<td align="left">2006-04-13</td>
<td align="left">original</td>
<td align="left">NA</td>
<td align="left">2006-06-19</td>
<td align="left">2015-09-16</td>
<td align="right">-0.18</td>
<td align="right">-9.43</td>
</tr>
</tbody>
</table>
<p>Examination of the ages of the parents reveals the issues being
reported.</p>
<p>We can remove the errors by changing the birth dates of <em>o4</em>
from 2006-04-13 to 2015-09-16 and of <em>d2</em> from 2015-09-16 to
2006-04-13.</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pedOne</span><span class="op">$</span><span class="va">birth_date</span><span class="op">[</span><span class="va">pedOne</span><span class="op">$</span><span class="va">ego_id</span> <span class="op">==</span> <span class="st">"o4"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.Date.html" class="external-link">as.Date</a></span><span class="op">(</span><span class="st">"2015-09-16"</span><span class="op">)</span></span>
<span><span class="va">pedOne</span><span class="op">$</span><span class="va">birth_date</span><span class="op">[</span><span class="va">pedOne</span><span class="op">$</span><span class="va">ego_id</span> <span class="op">==</span> <span class="st">"d2"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.Date.html" class="external-link">as.Date</a></span><span class="op">(</span><span class="st">"2006-04-13"</span><span class="op">)</span></span></code></pre></div>
<p>Note the changes made to the column names between the original
<strong>pedOne</strong> pedigree and the pedigree (<strong>ped</strong>)
we get from <strong>qcStudbook</strong>. We have chosen to limit the
displayed pedigree by selecting the <em>ego_id</em>’s and <em>id</em>’s
in <strong>pedOne</strong> and <strong>ped</strong> respectively.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ped</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qcStudbook.html">qcStudbook</a></span><span class="op">(</span><span class="va">pedOne</span>, minParentAge <span class="op">=</span> <span class="fl">0.0</span><span class="op">)</span></span>
<span><span class="va">ped</span><span class="op">[</span><span class="va">ped</span><span class="op">$</span><span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"s2"</span>, <span class="st">"d2"</span>, <span class="st">"o3"</span>, <span class="st">"o4"</span><span class="op">)</span>, <span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##   id sire  dam sex gen      birth exit  age recordStatus</span></span>
<span><span class="co">## 2 d2 &lt;NA&gt; &lt;NA&gt;   F   0 2006-04-13 &lt;NA&gt; 19.0     original</span></span>
<span><span class="co">## 4 s2 &lt;NA&gt; &lt;NA&gt;   M   0 2006-06-19 &lt;NA&gt; 18.8     original</span></span>
<span><span class="co">## 7 o3   s2   d2   F   1 2012-04-11 &lt;NA&gt; 13.0     original</span></span>
<span><span class="co">## 8 o4   s2   d2   M   1 2015-09-16 &lt;NA&gt;  9.6     original</span></span></code></pre>
<p>However, the preferred method of creating the standardized studbook
format with <strong>qcStudbook</strong> is to examine all errors found
and correcting them before proceeding. This is done by explicitly
requesting that all aspects inconsistent with the studbook format be
identified by setting <em>reportChanges</em> and <em>reportErrors</em>
to .</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">errorList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qcStudbook.html">qcStudbook</a></span><span class="op">(</span><span class="va">pedOne</span>,</span>
<span>  minParentAge <span class="op">=</span> <span class="fl">0.0</span>, reportChanges <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  reportErrors <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">errorList</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Error: The animal listed as a sire and also listed as a female is: s1.</span></span>
<span><span class="co">## Error: The animal listed as a dam and also listed as a male is: d1.</span></span>
<span><span class="co">## Change: The column where space was removed is: si re to sire.</span></span>
<span><span class="co">## Change: The columns where underscore was removed are: ego_id, dam_id, and birth_date to egoid, damid, and birthdate.</span></span>
<span><span class="co">## Change: The column changed from: egoid to id.</span></span>
<span><span class="co">## Change: The column changed from: damid to dam.</span></span>
<span><span class="co">## Change: The column changed from: birthdate to birth.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Please check and correct the pedigree file.</span></span>
<span><span class="co">## </span></span></code></pre>
<p>We will discuss each of these newly identified errors in a moment,
however, let’s look at shortening this report, because often you will
not be interested in the more trivial changes to the column names made
by <strong>qcStudbook</strong> and in those cases you simply choose not
to report changes to the column names as is shown here by setting
<em>reportChanges</em> to . For this illustration, we are going to bring
back the original copy of <strong>pedOne</strong> to see how the
suspicious parents are reported by the <strong>summary</strong>
function.</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pedOne</span> <span class="op">&lt;-</span> <span class="fu">nprcgenekeepr</span><span class="fu">::</span><span class="va"><a href="../reference/pedOne.html">pedOne</a></span></span>
<span><span class="va">errorList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qcStudbook.html">qcStudbook</a></span><span class="op">(</span><span class="va">pedOne</span>,</span>
<span>  minParentAge <span class="op">=</span> <span class="fl">0L</span>, reportChanges <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  reportErrors <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">90L</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">errorList</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Error: The animal listed as a sire and also listed as a female is: s1.</span></span>
<span><span class="co">## Error: The animal listed as a dam and also listed as a male is: d1.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Please check and correct the pedigree file.</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">## Animal records where parent records are suspicous because of dates.</span></span>
<span><span class="co">## One or more parents appear too young at time of birth.</span></span>
<span><span class="co">##   dam sire id sex      birth recordStatus exit  sireBirth   damBirth sireAge damAge</span></span>
<span><span class="co">## 2  d2   s1 o2   F 2009-03-17     original &lt;NA&gt; 2000-07-18 2015-09-16    8.66   -6.5</span></span>
<span><span class="co">## 3  d2   s2 o3   F 2012-04-11     original &lt;NA&gt; 2006-06-19 2015-09-16    5.81   -3.4</span></span>
<span><span class="co">## 4  d2   s2 o4   M 2006-04-13     original &lt;NA&gt; 2006-06-19 2015-09-16   -0.18   -9.4</span></span></code></pre>
<p>The first two errors mentioned are of particular interest. Currently
<strong>qcStudbook</strong> automatically changes the sex of dams to
<em>F</em> (female) and sires to <em>M</em> (male) when
<strong>reportErrors</strong> is set to .</p>
</div>
<div class="section level2">
<h2 id="genetic-loops">Genetic Loops<a class="anchor" aria-label="anchor" href="#genetic-loops"></a>
</h2>
<p>This feature is not supported within the Shiny application and is not
fully implemented.</p>
<p>To use the <strong>findLoops</strong> function run the following code
and select a pedigree as your input file that has a loop in it. We are
continuing to use the example pedigree that comes with the software
<em>Example_Pedigree.csv</em>.</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exampleTree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createPedTree.html">createPedTree</a></span><span class="op">(</span><span class="va">breederPed</span><span class="op">)</span></span>
<span><span class="va">exampleLoops</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findLoops.html">findLoops</a></span><span class="op">(</span><span class="va">exampleTree</span><span class="op">)</span></span></code></pre></div>
<p>You can count how many loops you have with the following code.</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">exampleLoops</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 3694</span></span></code></pre>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nLoops</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/countLoops.html">countLoops</a></span><span class="op">(</span><span class="va">exampleLoops</span>, <span class="va">exampleTree</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">nLoops</span><span class="op">[</span><span class="va">nLoops</span> <span class="op">&gt;</span> <span class="fl">0L</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 258</span></span></code></pre>
<p>You can list the first 10 sets of ids, sires and dams in loops with
the following line of code:</p>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">examplePedigree</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">exampleLoops</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"id"</span>, <span class="st">"sire"</span>, <span class="st">"dam"</span><span class="op">)</span><span class="op">]</span><span class="op">[</span><span class="fl">1L</span><span class="op">:</span><span class="fl">10L</span>, <span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##          id   sire    dam</span></span>
<span><span class="co">## 2519 V49H3Y UFI88T 9T7Y2Z</span></span>
<span><span class="co">## 2572 61FUGE UDQ5WC GL88CF</span></span>
<span><span class="co">## 2695 LWJ3A5 KZM9RB GCBYDW</span></span>
<span><span class="co">## 2722 RNQU14 H2RDE2 DKIM6U</span></span>
<span><span class="co">## 2752 L9M1DC 3PU50K WFQENR</span></span>
<span><span class="co">## 2755 Q8U9LB 3PU50K CLSVU6</span></span>
<span><span class="co">## 2905 FVJ14K UXC40T L5VC2M</span></span>
<span><span class="co">## 2922 531HAC UMV4BE 5DIPZN</span></span>
<span><span class="co">## 2924 85ESBB UQFY9C Q2RK1E</span></span>
<span><span class="co">## 2941 0VLW56 6KPKH7 MMEHXV</span></span></code></pre>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">elapsed_time</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_elapsed_time_str.html">get_elapsed_time_str</a></span><span class="op">(</span><span class="va">start_time</span><span class="op">)</span></span></code></pre></div>
<p>The current date and time is 2025-04-07 03:28:23.398707. The
processing time for this document was 13 seconds..</p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.3 (2025-02-28)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Random number generation:</span></span>
<span><span class="co">##  RNG:     Mersenne-Twister </span></span>
<span><span class="co">##  Normal:  Inversion </span></span>
<span><span class="co">##  Sample:  Rounding </span></span>
<span><span class="co">##  </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] nprcgenekeepr_1.0.7 knitr_1.50          ggplot2_3.5.1       stringi_1.8.7      </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] gtable_0.3.6         anytime_0.3.11       xfun_0.52            bslib_0.9.0         </span></span>
<span><span class="co">##  [5] htmlwidgets_1.6.4    lattice_0.22-6       vctrs_0.6.5          tools_4.4.3         </span></span>
<span><span class="co">##  [9] generics_0.1.3       tibble_3.2.1         pkgconfig_2.0.3      Matrix_1.7-2        </span></span>
<span><span class="co">## [13] data.table_1.17.0    checkmate_2.3.2      desc_1.4.3           readxl_1.4.5        </span></span>
<span><span class="co">## [17] lifecycle_1.0.4      farver_2.1.2         compiler_4.4.3       stringr_1.5.1       </span></span>
<span><span class="co">## [21] textshaping_1.0.0    munsell_0.5.1        Rlabkey_3.4.2        httpuv_1.6.15       </span></span>
<span><span class="co">## [25] htmltools_0.5.8.1    sass_0.4.9           yaml_2.3.10          htmlTable_2.4.3     </span></span>
<span><span class="co">## [29] later_1.4.1          pillar_1.10.2        pkgdown_2.1.1        jquerylib_0.1.4     </span></span>
<span><span class="co">## [33] cachem_1.1.0         sessioninfo_1.2.3    mime_0.13            tidyselect_1.2.1    </span></span>
<span><span class="co">## [37] digest_0.6.37        dplyr_1.1.4          labeling_0.4.3       WriteXLS_6.7.0      </span></span>
<span><span class="co">## [41] fastmap_1.2.0        grid_4.4.3           colorspace_2.1-1     cli_3.6.4           </span></span>
<span><span class="co">## [45] magrittr_2.0.3       withr_3.0.2          scales_1.3.0         promises_1.3.2      </span></span>
<span><span class="co">## [49] backports_1.5.0      plotrix_3.8-4        lubridate_1.9.4      timechange_0.3.0    </span></span>
<span><span class="co">## [53] rmarkdown_2.29       lambda.r_1.2.4       httr_1.4.7           futile.logger_1.4.3 </span></span>
<span><span class="co">## [57] cellranger_1.1.0     ragg_1.3.3           shiny_1.10.0         evaluate_1.0.3      </span></span>
<span><span class="co">## [61] rlang_1.1.5          futile.options_1.0.1 Rcpp_1.0.14          xtable_1.8-4        </span></span>
<span><span class="co">## [65] glue_1.8.0           formatR_1.14         rstudioapi_0.17.1    jsonlite_2.0.0      </span></span>
<span><span class="co">## [69] R6_2.6.1             systemfonts_1.2.2    fs_1.6.5</span></span></code></pre>
</div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Michael Raboin, Terry Therneau, Amanda Vinson, R. Mark Sharp, Matthew Schultz, Southwest National Primate Research Center NIH grant P51 RR13986, Oregon National Primate Research Center grant P51 OD011092.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
