<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Genetic Management Tools Manual â€¢ nprcgenekeepr</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Genetic Management Tools Manual">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">nprcgenekeepr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.7</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/a2interactive.html">Interactive Use of nprcgenekeepr</a></li>
    <li><a class="dropdown-item" href="../articles/a3manual.html">Genetic Management Tools Manual</a></li>
    <li><a class="dropdown-item" href="../articles/ColonyManagerTutorial.html">GeneKeepR: A Colony Manager's Tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/simulatedKValues.html">Quick Example of Simulated Kinships with Partial Parentage</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/rmsharp/nprcgenekeepr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Genetic Management Tools Manual</h1>
                        <h4 data-toc-skip class="author">Michael Raboin,
Amanda Vinson, and R. Mark Sharp</h4>
            
            <h4 data-toc-skip class="date">April 12, 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rmsharp/nprcgenekeepr/blob/master/vignettes/a3manual.Rmd"><code>vignettes/a3manual.Rmd</code></a></small>
      <div class="d-none name"><code>a3manual.Rmd</code></div>
    </div>

    
    
<p><a href="#introduction">Introduction</a><br><a href="#summary-of-major-functions">Summary of Major
Functions</a><br><a href="#installation">Installation</a><br><a href="#running-shiny-application">Running Shiny Application</a><br><a href="#input">Input</a><br><a href="#pedigree-browser">Pedigree Browser</a><br><a href="#genetic-value-analysis">Genetic Value Analysis</a><br><a href="#summary-statistics">Summary Statistics</a><br><a href="#breeding-group-formation">Breeding Group Formation</a><br><a href="#orip-reporting">ORIP Reporting</a><br><a href="#algorithm-breeding-group-formation">Algorithm: Breeding Group
Formation</a><br><a href="#algorithm-genome-uniqueness">Algorithm: Genome
Uniqueness</a></p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The goal of <strong>nprcgenekeepr</strong> is to implement Genetic
Tools for Colony Management. It was initially conceived and developed as
a Shiny web application at the Oregon National Primate Research Center
(ONPRC) to facilitate some of the analyses they perform regularly. It
has been enhanced to have more capability as a Shiny application and to
expose the functions so they can be used either interactively or in R
scripts.</p>
<p>This work has been supported in part by NIH grants P51 RR13986 to the
Southwest National Primate Research Center and P51 OD011092 to the
Oregon National Primate Research Center.</p>
<!--It is now managed and maintained as a joint effort between ONPRC-->
<!--and Southwest National Primate Research Center (SNPRC) with the -->
<!--coding being done by R. Mark Sharp, Ph.D.-->
<p>At present, the application supports 5 functions:</p>
<ol style="list-style-type: decimal">
<li>Quality control of studbooks contained in text files or Excel
workbooks and of pedigrees within LabKey Electronic Health Records
(EHR)</li>
<li>Creation of pedigrees from a lists of animals using the LabKey EHR
integration</li>
<li>Creation and display of an age by sex pyramid plot of the living
animals within the designated pedigree</li>
<li>Generation of Genetic Value Analysis Reports</li>
<li>Creation of potential breeding groups with and without proscribed
sex ratios and defined maximum kinships.</li>
</ol>
<p><strong>For more information see:</strong><br>
A Practical Approach for Designing Breeding Groups to Maximize Genetic
Diversity in a Large Colony of Captive Rhesus Macaques (<em>Macaca
mulatto</em>) Vinson, A ; Raboin, MJ <em>Journal Of The American
Association For Laboratory Animal Science</em>, 2015 Nov, Vol.54(6),
pp.700-707 [Peer Reviewed Journal]</p>
</div>
<div class="section level2">
<h2 id="summary-of-major-functions">Summary of Major Functions<a class="anchor" aria-label="anchor" href="#summary-of-major-functions"></a>
</h2>
<div class="section level3">
<h3 id="quality-control">Quality Control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h3>
<p>Studbooks maintained by breeding colonies generally contain
information of varying quality. The quality control functions of the
toolkit check to ensure all animals listed as parents have their own
line entries, all parents have the appropriate sex listed, no animals
are listed as both a sire and a dam, duplicate entries are removed,
pedigree generation numbers are added, and all dates are valid dates. In
addition, exit dates are added if possible and are consistent with other
information such as departure dates and death dates. Current ages of
animals that are still alive are added if a database connection is
provided via a configuration file and the user has read permission on a
LabKey server with the demographic data in an <em>EHR</em> (Electronic
Health Record) module. See <a href="https://www.labkey.org/Documentation/wiki-page.view?name=netrc" class="external-link">LabKey
documentation</a>.</p>
<p>Parents with ages below a user selected threshold are identified. A
minimum parent age in years is set by the user and is used to ensure
each parent is at least that age on the birth date of an offspring. The
minimum parent age defaults to 2 years. This check is not performed for
animals with missing birth dates.</p>
</div>
<div class="section level3">
<h3 id="creation-of-pedigree-from-a-list-of-potential-breeders-and-labkey">Creation of Pedigree From a List of Potential Breeders and
LabKey<br><a class="anchor" aria-label="anchor" href="#creation-of-pedigree-from-a-list-of-potential-breeders-and-labkey"></a>
</h3>
<p>The user can enter a list of focal animals in a CSV file that will be
used to create a pedigree containing all direct relative (ancestors and
descendants) via the <strong>labkey.selectRows</strong> function within
the <strong>Rlabkey</strong> package if a database connection is
provided via a configuration file and the user has read permission on a
LabKey server with the demographic data in an <strong>EHR</strong>
(Electronic Health Record) module.</p>
<p>Two configuration files are needed to use the database features of
nprcgenekeepr with LabKey. The first file is named
<strong>_netrc</strong> on Microsoft Windows operating systems and
<strong>.netrc</strong> otherwise, allows the user to authenticate with
LabKey through the LabKey API and is fully described by <a href="https://www.labkey.org/Documentation/wiki-page.view?name=netrc" class="external-link">LabKey
documentation</a></p>
<p>The second file is named <strong>_nprcgenekeepr_config</strong> on
Microsoft Windows operating systems and
<strong>.nprcgenekeepr_config</strong> otherwise and is the
<code>nprcgenekeepr</code> <a href="https://github.com/rmsharp/nprcgenekeepr/blob/master/inst/extdata/example_nprcgenekeepr_config">configuration
file</a> An image of this example configuration file is included as a
data object and can be loaded and viewed with the following lines of R
code in the R console.</p>
</div>
<div class="section level3">
<h3 id="display-of-an-age-by-sex-pyramid-plot">Display of an age by sex pyramid plot<a class="anchor" aria-label="anchor" href="#display-of-an-age-by-sex-pyramid-plot"></a>
</h3>
<p>Adapted from <a href="https://www.thoughtco.com/age-sex-pyramids-and-population-pyramids-1435272" class="external-link">https://www.thoughtco.com/age-sex-pyramids-and-population-pyramids-1435272</a>
on 20190603. Written by Matt Rosenberg. Updated May 07, 2019.</p>
<p>The most important demographic characteristic of a population is its
age-sex structure. Age-sex pyramids (also known as population pyramids)
graphically display this information to improve understanding and make
comparison easy. The population pyramid sometimes has a distinctive
pyramid-like shape when displaying a growing population.</p>
<div class="section level4">
<h4 id="how-to-read-the-age-sex-graph">How to Read the Age-Sex Graph<a class="anchor" aria-label="anchor" href="#how-to-read-the-age-sex-graph"></a>
</h4>
<p>An age-sex pyramid breaks down a population into male and female
genders and age ranges. Usually, youâ€™ll find the left side of the
pyramid graphing the male population and the right side of the pyramid
displaying the female population.</p>
<p>Along the horizontal axis (x-axis) of a population pyramid, the graph
displays the population either as a total population of that age or as a
percentage of the population at that age. The center of the pyramid
starts at zero population and extends out to the left for males and
right for females in increasing size, or proportion of the
population.</p>
<p>Along the vertical axis (y-axis), age-sex pyramids display two-year
age increments, from birth at the bottom to old age at the top.</p>
</div>
</div>
<div class="section level3">
<h3 id="genetic-value-analysis-reports">Genetic Value Analysis Reports<a class="anchor" aria-label="anchor" href="#genetic-value-analysis-reports"></a>
</h3>
<p>The Genetic Value Analysis is a ranking scheme developed at ONPRC to
indicate the relative breeding value of animals in the colony. The
scheme uses the mean kinship for each animal to indicate how
inter-related it is with the rest of the current breeding colony
members. Genome uniqueness is used to provide an indication of whether
or not an animal is likely to possess alleles at risk of being lost from
the colony. Under the scheme, animals with low mean kinship or high
genome uniqueness are ranked more highly.</p>
</div>
<div class="section level3">
<h3 id="breeding-group-formation">Breeding Group Formation<a class="anchor" aria-label="anchor" href="#breeding-group-formation"></a>
</h3>
<p>One of the goals in breeding group formation is to avoid the
potential for mating of closely related animals. Since behavioral
concerns and housing constraints will also be taken into account in the
group formation process, it is our goal to provide the largest number of
animals possible from a list of candidates that can be housed together
without risk of consanguineous mating. To that end, this function uses
information from the Genetic Value Analysis to search for the largest
combinations of animals that can be produced from a list of
candidates.</p>
<p>The default options do not consider the sex of individuals when
forming the groups, though this has likely been a consideration by the
user in selecting the candidate group members. Optionally the user may
select to form harem groups, which considers the sex of individuals when
forming groups and restricts the number of males to one per group.</p>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the CRAN version of <strong>nprcgenekeepr</strong>
from the R console prompt with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"rmsharp"</span>, <span class="st">"nprcgenekeepr"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>You can install the development version of
<strong>nprcgenekeepr</strong> from GitHub from the R console prompt
with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"rmsharp"</span>, <span class="st">"nprcgenekeepr"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>All missing dependencies should be automatically installed.</p>
</div>
<div class="section level2">
<h2 id="running-shiny-application">Running Shiny Application<a class="anchor" aria-label="anchor" href="#running-shiny-application"></a>
</h2>
<p>The toolset available within nprcgenekeepr can be used inside
standard R scripts. However, it was originally designed to be used
within a Shiny application that can be started with:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rmsharp.github.io/nprcgenekeepr/" class="external-link">nprcgenekeepr</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/runGeneKeepR.html">runGeneKeepR</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="input">Input<a class="anchor" aria-label="anchor" href="#input"></a>
</h2>
<p>The Input tab is the starting point for all analyses. The file should
be a delimited, regular text file with a header row specifying the
columns. The tab provides information on the allowable columns in input
files, and how the columns will be used in quality control of the data.
Quality control of studbook data occurs automatically upon file
upload.</p>
<p>Presently, the only columns required are those specifying the Ego ID,
Sire ID, Dam ID, and Sex. The remaining columns listed are optional, but
will be used if they are present in the uploaded file. The table of the
tab describes how these optional columns will be used. Additionally, the
panel on the left of this tab provides options that can be used during
the upload and QC process, such as specifying the field separator used
in the uploaded file.</p>
<p>During quality control, a flag is added for the current, living
population. This flag is generated based on the information columns
provided and is fairly specific to how the breeding population is
defined at ONPRC. Two of the options for specifying the population of
interest can be toggled through this panel, however. Normally, the
breeding colony is restricted to Indian-origin, SPF 4 animals. These two
restrictions can be turned off by setting the options on this panel.</p>
<p>Additionally, the population of interest can be specified directly in
either the input file, or entered on the Pedigree Browser tab.</p>
</div>
<div class="section level2">
<h2 id="pedigree-browser">Pedigree Browser<a class="anchor" aria-label="anchor" href="#pedigree-browser"></a>
</h2>
<p>The Pedigree Browser tab allows the user to view the input data,
specify a population to examine, and output the cleaned studbook or
trimmed pedigree.</p>
<p>Upon uploading a studbook file, the data goes through the quality
control process described on the Input File Format tab. The cleaned
version is displayed on this tab. By default, the entire uploaded
studbook will be available for viewing on this tab. The first 10 rows
will automatically be shown, but this range can be adjusted using the
input boxes at the top. The default setting of showing only the first 10
rows is due to the size of the full ONPRC studbook: loading all 32,000
animals can cause the application to be slow.</p>
<p>The tab also contains functionality for trimming the pedigree. By
checking the provided box, the studbook that was uploaded will be
trimmed down to just the ancestors of the currently specified
population. This will remove any lineages that havenâ€™t contributed to
the focal group.</p>
<p>As stated above, this tab will also allow the user to directly
specify a population to examine. In the top half of the tab, there is an
input box to specify the focal set of animals. The population flag can
be reset by adding the desired animal IDs to the box. Once the
population flag has been set to the desired group of animals, all
further analyses will be relative to this group.</p>
</div>
<div class="section level2">
<h2 id="genetic-value-analysis">Genetic Value Analysis<a class="anchor" aria-label="anchor" href="#genetic-value-analysis"></a>
</h2>
<p>The Genetic Value Analysis tab provides all of the options needed for
producing a genetic value analysis report. The specifics of generating
the genetic value analysis report are described on the
<strong><em>Genetic Value Analysis and Breeding Group
Description</em></strong> tab.</p>
<p>When the analysis is begun, it will generate a genetic value analysis
for the currently-specified population in the pedigree. If no population
has been specified, the entire pedigree will go into the analysis. This
can be problematic, as the function for calculating the pairwise kinship
matrix cannot handle large pedigrees. The kinship calculation is known
to handle pedigree files containing up to 6000 individuals. It will not,
however, handle the whole ONPRC rhesus studbook (~24,000 animals). The
exact maximum pedigree size is not currently known and will need to be
tested. Due to these problems, the input studbook will automatically be
trimmed to the ancestors of the currently-specified population before
the genetic value analysis is begun.</p>
<p>The genome uniqueness threshold input box allows the user to specify
what constitutes a â€˜uniqueâ€™ allele in the gene-drop simulation. The
algorithm description later in this document provides a more in-depth
explanation of how the genome uniqueness calculation uses this
information. By default, the gene-drop simulation underlying the genome
uniqueness calculation considers an individual as unique if no other
members of the current population have inherited the same allele during
an iteration of the gene-drop. This can be adjusted using the drop-down
box to allow up to four other animals to have inherited the allele and
still consider it unique.</p>
<p>After the report has been generated, it can be subset to view a
specific group of the animals using the text input box. Both the
currently-viewed subset and the full report can be exported to a file
from here.</p>
</div>
<div class="section level2">
<h2 id="summary-statistics">Summary Statistics<a class="anchor" aria-label="anchor" href="#summary-statistics"></a>
</h2>
<p>This tab provides some descriptions of the population being examined
after the genetic value analysis has been run. The tab reports the
number of known founders, female founders, male founders, founder
equivalents, and founder genome equivalents with the first table, which
has a single row. The second table has a row for Mean Kinship and a row
for Genome Uniqueness. Each row has the Tukey five number summary, which
is the minimum, 1<sup>st</sup> quartile, mean, median, 3<sup>rd</sup>
quartile, and maximum. Lastly, the tab displays histograms and box plots
of the distribution of mean kinship coefficients, the distribution of
mean kinship coefficient Z-scores, Distribution of genome uniqueness
values.</p>
</div>
<div class="section level2">
<h2 id="breeding-group-formation-1">Breeding Group Formation<a class="anchor" aria-label="anchor" href="#breeding-group-formation-1"></a>
</h2>
<p>The last major function of the R-package is to aid in generating
breeding groups that avoid inter-animal relatedness. This tab allows you
to build a number of breeding groups from a specified list of candidate
animals. It also has an option to build a group by adding animals from a
list of candidates to a currently-existing group.</p>
<p>In the top half of the tab, there are entry boxes and menus to adjust
the options of the analysis. By default, the analysis will ignore
relatedness between animals that is more distant than the second cousin
level, pairwise relatedness involving an animal under 1 year of age, and
relatedness between all females. All of these options can be adjusted
before the analysis is run, however.</p>
<p>If the desire is to add animals to an existing group, the IDs of the
candidate animals can be entered into the first text box (just as they
would be if a new group were being generated). The IDs of the current
group members can be added into the second text box. It should be noted
that it will cause an error if the provided candidate animals or current
group member IDs are not part of the population for which a genetic
value analysis was run. The kinship matrix produced for this analysis
provides the pairwise kinship values used by the group formation
functions.</p>
<p>After the simulation is done, the first group will be displayed
automatically. The group being displayed can be changed with the
drop-down menu. Whichever group is currently being displayed can then be
downloaded with the export button.</p>
</div>
<div class="section level2">
<h2 id="genetic-value-analysis-and-breeding-group-formation-description">Genetic Value Analysis and Breeding Group Formation Description<a class="anchor" aria-label="anchor" href="#genetic-value-analysis-and-breeding-group-formation-description"></a>
</h2>
<p>This tab contains more in-depth descriptions of how the Genetic Value
Analysis is created, and how breeding groups are formed by the
program.</p>
</div>
<div class="section level2">
<h2 id="orip-reporting">ORIP Reporting<a class="anchor" aria-label="anchor" href="#orip-reporting"></a>
</h2>
<p>The ORIP Reporting tab will eventually contain information for
reporting to the Office of Research Infrastructure Programs (ORIP). This
tab may end up being merged with the Summary Statistics tab and contain
a number of statistics, tables and histograms. Alternatively, this may
contain a subset of information from the Summary Statistics tab
presented as a formatted report that can be exported and submitted to
ORIP. The exact information that needs to be submitted for ORIP
recordkeeping is still under discussion.</p>
</div>
<div class="section level2">
<h2 id="algorithm-breeding-group-formation">Algorithm: Breeding Group Formation<a class="anchor" aria-label="anchor" href="#algorithm-breeding-group-formation"></a>
</h2>
<p>The group formation process is accomplished by using an algorithm for
determining the maximal independent set (MIS). In graph theory, a
maximal independent set is the largest set of vertices in a graph where
no two share an edge. In breeding group formation, the vertices are
animals, and the edges are the kinships that need to be considered. For
a given group of animals and pairwise kinships, there are potentially
many maximal independent sets, depending on which animals are included
or excluded from the final group. In order to effectively sample the set
of MISs, we use random selection of animals and repeat the MIS
generation numerous times. This allows us to sample a number of MISs and
then choose the one that best fits our selection criteria. For our
purposes, we want the largest group that can be formed from this set of
animals, where none have concerning relatedness to each other.</p>
<p>The algorithm requires several pieces of information:</p>
<ol style="list-style-type: decimal">
<li>The candidate animals<br>
</li>
<li>A matrix of pairwise kinships between candidate animals<br>
</li>
<li>The number of groups desired from the list of candidate
animals<br>
</li>
<li>The number of simulations to run.<br>
* This is equivalent to the number of random MISs to generate and
compare.<br>
</li>
<li>Information on which inter-animal relationships (if any) should be
ignored.</li>
</ol>
<div class="section level4">
<h4 id="data-pre-processing">Data Pre-processing<a class="anchor" aria-label="anchor" href="#data-pre-processing"></a>
</h4>
<p>Before the group formation algorithm begins generating MISs, the data
is pre-processed to remove any animals and pairwise kinships that should
not be considered.</p>
<p>Specifically:</p>
<ol style="list-style-type: decimal">
<li>The candidate animals provided are checked, and any that were
designated as low-value by the genetic value analysis will be removed
from further consideration.<br>
* This behavior can be toggled off to allow low-value animals in the
formation process<br>
</li>
<li>The pairwise kinship data is filtered down to only the kinship
between candidate animals.<br>
</li>
<li>If an age threshold has been set, kinships involving animals below
the threshold will be filtered out.<br>
* This allows the algorithm to ignore young animals, as young animals
typically go to whatever social group their dam does.<br>
* By default, we ignore animals under 1 year of age<br>
</li>
<li>Pairwise kinships below the specified level will be filtered
out.<br>
* By default, we ignore relatedness more distant than 2nd cousin<br>
</li>
<li>Pairwise kinships between females will be filtered out<br>
* This allows females of the same matriline to be part of the same group
like they would be in the wild.<br>
* This behavior can be toggled off to prevent relatedness between
females.</li>
</ol>
</div>
<div class="section level4">
<h4 id="random-maximum-independent-set-generation">Random Maximum Independent Set Generation<a class="anchor" aria-label="anchor" href="#random-maximum-independent-set-generation"></a>
</h4>
<p>After any animals and relationships that should be ignored are
removed from the dataset, the algorithm begins using the remaining
animals and kinship information to generate potential groups.</p>
<p>The algorithm proceeds by the following steps:</p>
<ol style="list-style-type: decimal">
<li>For <strong>I</strong> iterations:
<ol style="list-style-type: lower-alpha">
<li>Generate <strong>N</strong> empty sets, where <strong>N</strong> is
the desired number of groups to be created.<br>
</li>
<li>While there are candidate animals remaining:<br>
i. Pick an animal <strong>A</strong> randomly from the set of candidate
animals<br>
ii. Choose a group <strong>G</strong> randomly from one of the
<strong>N</strong> groups, and assign <strong>A</strong> to it<br>
iii. Remove animal <strong>A</strong> from consideration for all
<strong>N</strong> groups<br>
iv. Remove all animals related to <strong>A</strong> from consideration
from for group <strong>G</strong><br>
</li>
<li>Score the groups that were generated<br>
i. For our purposes, we calculate the average group size<br>
</li>
<li>If the score of the new groups is higher than groups that were
previously generated, save the new groups.<br>
</li>
</ol>
</li>
<li>Return the currently saved groups
<ol style="list-style-type: lower-alpha">
<li>This should be the best groups encountered in <strong>I</strong>
iterations.</li>
</ol>
</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="algorithm-genome-uniqueness">Algorithm: Genome Uniqueness<a class="anchor" aria-label="anchor" href="#algorithm-genome-uniqueness"></a>
</h2>
<p>Genome uniqueness is calculated through the use of a gene-drop
simulation to estimate how frequently an animal will possess founder
alleles not present in other members of the focal population, or present
in a specified number or fewer.</p>
<p>The gene-drop simulation used by the web application is a vectorized
version and is shown in the figure below. In an un-vectorized version,
if 5000 gene-drop simulations are desired for the estimation process,
the population had to be iterated over 5000 times. Since each iteration
of the gene-drop is independent, the process can be vectorized so that
each element of a vector represents 1 iteration of the gene-drop
simulation. In the vectorized version, the population is iterated over
once, regardless of the number of simulations desired. This drastically
reduces the amount of time necessary for the program to run.</p>
<div class="section level4">
<h4 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h4>
<p>The basic steps of the gene-drop are:</p>
<ol style="list-style-type: decimal">
<li>Each founder is assigned two unique alleles<br>
</li>
<li>For each subsequent generation:
<ol style="list-style-type: lower-alpha">
<li>Assign genotypes to each member of the generation
<ul>
<li>For each animal, find the genotypes of the parents, and select<br>
one allele from each parent randomly.</li>
</ul>
</li>
</ol>
</li>
</ol>
<p>Once every animal has been assigned a genotype by mendelian
inheritance tally the number of unique alleles possessed by each member
of the focal population. In the case of this algorithm, we do allow the
â€˜uniquenessâ€™ threshold to be adjusted so that an allele can be
considered unique if it is possessed by N or fewer other members of the
focal population.</p>
</div>
<div class="section level4">
<h4 id="vectorized-gene-drop-details">Vectorized Gene-Drop Details<a class="anchor" aria-label="anchor" href="#vectorized-gene-drop-details"></a>
</h4>
<p>The vectorized gene-drop simulation follows the same basic process
described above. The difference is that instead of dropping one allele
at a time, and repeating the simulation N times, the vectorized version
drops N independent alleles one time.</p>
<p>In the vectorized version, each animal has a vector of paternally
inherited alleles and a vector of maternally inherited alleles. For each
offspring, a random combination of these alleles is produced and dropped
down to the offspring by the process below and shown in the following
figure:</p>
<ol style="list-style-type: decimal">
<li>To start the simulation, each founder is assigned two unique
founding alleles.<br>
N-element vectors are created of these alleles, where N is the desired
number of<br>
simulations. In the example below, this founder was assigned the unique
founder<br>
alleles 1 &amp; 2 and 5 simulations were desired.<br>
</li>
<li>Each time alleles need to be dropped from parent to offspring, a
unique<br>
transmission vector is created representing whether or not an
allele<br>
was passed to that offspring. The vector is generated to contain a
random combination<br>
of 0â€™s and 1â€™s. The animalâ€™s paternally inherited alleles are then
multiplied by the<br>
transmission vector, while the maternally inherited alleles are
multiplied by the<br>
compliment of the transmission vector.<br>
</li>
<li>To generate the final set of alleles received by the offspring, the
maternal and<br>
paternal allele vectors are added together.<br>
</li>
<li>The result is a vector of alleles that this offspring has received
from this parent.</li>
</ol>
<p>Once allele vectors have been generated for every animal in the
pedigree, the focal population can be subset out. Within this population
of allele vectors, unique alleles can be determined:</p>
<p>For each position on the allele vectors (1:N) - Gather each animalâ€™s
two alleles - If the number of other animals possessing that allele is
equal to, or below the threshold, score the allele as unique (1) -
Otherwise, score the allele as non-unique (0)</p>
<p>Once every position on each animalâ€™s two allele vectorâ€™s has been
scored, sum all of the scores for an animal and divide by the total
number of alleles being considered (2 * number of simulations).</p>
<div class="float">
<img src="../reference/figures/GeneDrop.png" alt="Generation of a vector of five gametes from one parent. Showing how the transmission vectors (row 2) determine which alleles are passed from the parental alleles or haplotypes (row 1) to form complementary vectors (row 3) that are combined by adding corresponding elements to form the final vector of transmitted alleles (row 4)."><div class="figcaption">Generation of a vector of five gametes from one
parent. Showing how the transmission vectors (row 2) determine which
alleles are passed from the parental alleles or haplotypes (row 1) to
form complementary vectors (row 3) that are combined by adding
corresponding elements to form the final vector of transmitted alleles
(row 4).</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="software-issues">Software Issues<a class="anchor" aria-label="anchor" href="#software-issues"></a>
</h2>
<p>Our goal is to use current R software development practices in an
open software environment. Users can see all of the code at <a href="https://github.com/rmsharp/nprcgenekeepr">github.com/rmsharp/nprcgenekeepr</a>
and can submit suggestions and bug reports on our issue tracker at <a href="https://github.com/rmsharp/nprcgenekeepr/issues">github.com/rmsharp/nprcgenekeepr/issues</a>.</p>
<div class="section level3">
<h3 id="cicd-pipeline-use">CICD Pipeline Use<a class="anchor" aria-label="anchor" href="#cicd-pipeline-use"></a>
</h3>
<p>The application and associated website is being continuously
integrated at each push to the online repository. While often new
features being added are not stable or complete, it is uncommon for the
application not to run and perform functions that were working before.
However, make sure the build was passing by looking for a green
<em>R-CMD-check.yaml Passing</em> badge at the top of the README file at
<a href="https://github.com/rmsharp/nprcgenekeepr/">https://github.com/rmsharp/nprcgenekeepr/</a>.</p>
</div>
<div class="section level3">
<h3 id="debug-logging">Debug Logging<a class="anchor" aria-label="anchor" href="#debug-logging"></a>
</h3>
<p>There is a logging system integrated into the package using the
package <strong>futile.logger</strong>. Note the checkbox at the bottom
of the side panel on the <em>Input</em> tab. When the <em>Debug on</em>
checkbox is checked (it is not checked by default), the application
writes to a file named <em>nprcgenekeepr.log</em> in the users home
directory. Currently, events occurring the the <em>server.R</em> file
are logged as that is where most errors are exposed.</p>
</div>
<div class="section level3">
<h3 id="code-coverage">Code Coverage<a class="anchor" aria-label="anchor" href="#code-coverage"></a>
</h3>
<p>Code coverage reports are part of the automated build system running
in GitHub Actions. We are using the <strong>testthat</strong> package
for unit tests. Currently all code returning values that do not access a
database or the file system have coverage with unit tests. Many of these
have 100 percent of the lines covered. However, the unit tests are not
exhaustive. The practice is to add further tests as errors are detected
or when working on the code and a new unit test possibility is
discovered. As of 20241223 95.70 percent of the lines are covered.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Michael Raboin, Terry Therneau, Amanda Vinson, R. Mark Sharp, Matthew Schultz, Southwest National Primate Research Center NIH grant P51 RR13986, Oregon National Primate Research Center grant P51 OD011092.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
